```{r, include = FALSE, eval = F}
knitr::opts_knit$set(out.format = "html")
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%'
)
options(width=100)

gen_figs <- F
figpath <- "../inst/cdoc/RAPToR-refbuilding_figs/"
if(!file.exists(figpath)){
  dir.create(figpath)
}


library(RAPToR)
library(RColorBrewer)
library(beeswarm)

requireNamespace("limma", quietly = T)
requireNamespace("stats")

transp <- function(col, a=.5){
  colr <- col2rgb(col)
  return(rgb(colr[1,], colr[2,], colr[3,], a*255, maxColorValue = 255))
}

png_custom <- function(figname, path = "", 
                       fig.width = 7, fig.height = 5, res = 150, ...){
  png(filename = paste0(path, figname, ".png"), 
      width = fig.width, height = fig.height, res = res, units = "in")
}

show_fig <- function(figname = knitr::opts_current$get("label"), expr, path = figpath, ...){
  if(gen_figs){
    png_custom(figname = figname, path = figpath, ...)
    eval(expr = expr)
    dev.off()
  }
  else{
    knitr::include_graphics(paste0(path, figname, ".png"))
  }
}
```

### The data

Here, we use two *Drosophila melanogaster* embryo development time-series. 

1. An embryo development time-series, part of the modENCODE project and published by @graveley2011developmental, hereafter called `dsgraveley2011`. This is the data used to build the reference. (Data downloaded from [fruitfly.org](https://fruitfly.org/sequence/download.html))
 1. A high-resolution time-series of single embryos published by @levin2016mid, called `dslevin2016dmel`. This data is used for external validation. (Accession : [GSE60471](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60471))

Of note, the reference data has a low time resolution, which displays the effectiveness of gene expression interpolation.

#### {.tabset}
Code to generate `dsgraveley2011` and `dslevin2016dmel` :

##### Hide

##### Show
**Note : set the `data_folder` variable to an existing path on your system where you want to store the objects.**

```{r ex2_load_namespaces, eval = F}
data_folder <- "../inst/extdata/"

requireNamespace("utils", quietly = T)
requireNamespace("GEOquery", quietly = T) # May need to be installed with bioconductor
requireNamespace("Biobase", quietly = T)
```


```{r ex2_c2tpm_func, code = readLines("convert2tpm.R"), echo = TRUE, eval=FALSE}
```


```{r ex2_load_droso_genes, code = readLines("load_droso_genes.R"), echo = TRUE, eval=FALSE}
```

<br>

###### `dsgraveley2011`

```{r ex2_load_dsgraveley2011_c, code = readLines("load_dsgraveley2011.R"), echo = TRUE, eval=FALSE}
```


<br>

###### `dslevin2016dmel`
```{r ex2_load_dslevin2016dmel_c, code = readLines("load_dslevin2016dmel.R"), echo = TRUE, eval=FALSE}
```

```{r cleanup_genes, eval = F}
rm(droso_genes, raw2tpm, fpkm2tpm)
```


#### Normalization & Quick look
```{r ex2_load, include=FALSE, eval=gen_figs}
load("../inst/extdata/dsgraveley2011.RData")
load("../inst/extdata/dslevin2016dmel.RData")
```

```{r ex2_qnorm_log, eval=gen_figs}
dsgraveley2011$g <- limma::normalizeBetweenArrays(dsgraveley2011$g, method = "quantile")
dsgraveley2011$g <- log1p(dsgraveley2011$g)

dslevin2016dmel$g <- limma::normalizeBetweenArrays(dslevin2016dmel$g, method = "quantile")
dslevin2016dmel$g <- log1p(dslevin2016dmel$g)
```

```{r ex2_ql_c1, results='markup', eval=gen_figs}
dsgraveley2011$g[1:5, 1:5]
#>               em0.2hr   em2.4hr   em4.6hr   em6.8hr  em8.10hr
#> FBgn0000003 3.9651391 4.2738527 3.3174101 4.5644242 4.6982706
#> FBgn0000008 1.2949845 0.9215699 0.6958672 0.6476801 0.7445991
#> FBgn0000014 0.5099295 0.9512866 1.3952815 1.8610406 1.8421960
#> FBgn0000015 0.2435639 0.6423988 1.0511912 1.1094674 1.0194280
#> FBgn0000017 1.7968429 2.0901351 1.3389420 1.5336183 1.6777064


head(dsgraveley2011$p, n = 5)
#>      sname age
#> 1  em0.2hr   0
#> 2  em2.4hr   2
#> 3  em4.6hr   4
#> 4  em6.8hr   6
#> 5 em8.10hr   8
```

##### Correlation Matrix
```{r ex2_ql_c2, echo = F, fig.height=5, fig.width=5, out.width="60%"}
show_fig(expr = {
  cor_dsgraveley2011 <- cor(dsgraveley2011$g, method = "spearman")
  ord <- order(dsgraveley2011$p$age)
  heatmap(cor_dsgraveley2011[ord, ord], Colv = NA, Rowv = NA, scale = "none", keep.dendro = F, margins = c(2,5),
          labRow = "", labCol = "")
  par(xpd = T)
  mtext(text = unique(dsgraveley2011$p$age), side = 1, line = 4, at = seq(-.16,.85, l = 12))
  
  # color key
  image(x = c(.95,1), y = seq(0.6,1, l = 9), useRaster = T,
        z = matrix(seq(min(cor_dsgraveley2011), max(cor_dsgraveley2011), l = 9), ncol = 9),
        col = hcl.colors(12, "YlOrRd", rev = TRUE), add = T)
  text(.975, 1, pos = 3, labels = expression(rho), font = 2)
  text(1, y = seq(0.6,1, l = 9)[c(T,F)], pos = 4, 
       labels = round(seq(min(cor_dsgraveley2011), max(cor_dsgraveley2011), l = 9)[c(T,F)], 2), cex = .6)
  
  
  xlp <- 1.025
  mtext(at = xlp, line = 4, side = 1, text = "(hours)", cex = .8)
}, fig.height=5, fig.width=5)
```

##### Plotting components

```{r ex2_ql_c3, eval=gen_figs}
pca_dsgraveley2011 <- stats::prcomp(t(dsgraveley2011$g), rank = 12,
                                    center = TRUE, scale = FALSE)
```

```{r ex2_ql_c4, echo = F, fig.height=6, fig.width=12}
show_fig(expr = {
  par(mfrow = c(2,4))
  invisible(sapply(seq_len(8), function(i){
    plot(dsgraveley2011$p$age, pca_dsgraveley2011$x[,i], lwd = 2,
         xlab = "age", ylab = "PC", main = paste0("PC", i))
  
    points(dsgraveley2011$p$age, pca_dsgraveley2011$x[,i], type = 'l', lty = 2)
  }))
}, fig.height=6, fig.width=12)
```


### Model fitting

#### Component number

```{r ex2_mf_c1, eval=gen_figs}
nc <- sum(summary(pca_dsgraveley2011)$importance[3,] < .99) + 1
nc
#> [1] 8
```

#### Model

```{r ex2_mf_c2, eval=gen_figs}
m_dsgraveley2011 <- ge_im(X = dsgraveley2011$g, p = dsgraveley2011$p, formula = "X ~ s(age, bs = 'cr')", nc = nc)
```

```{r ex2_mf_c3, echo = F, eval=gen_figs}
mp_dsgraveley2011 <- mperf(dsgraveley2011$g, predict(m_dsgraveley2011), is.t = T)
as.data.frame(mp_dsgraveley2011, row.names = "")
```

```{r ex2_mf_print}
#>        aCC       aRE         MSE      aRMSE
#>  0.9400274  0.4040618 0.009908293 0.09954041
```


#### Validation

##### Build ref object


```{r ex2_vd_c1, results='markup', eval = gen_figs}
r_dsgraveley2011 <- make_ref(m = m_dsgraveley2011,
                             n.inter = 100,                    # interpolation resolution
                             t.unit = "h past egg-laying",     # time unit
                             metadata = list("organism" = "D. melanogaster", # any metadata
                                             "profiling" = "whole-organism, bulk",
                                             "technology" = "RNAseq")) 
```

##### Plot component predictions

```{r ex2_vd_c2, echo = F, fig.width=12, fig.height=6}
show_fig(expr = {
  
  par(mfrow = c(2,4))
  plot(m_dsgraveley2011, r_dsgraveley2011, col.i = 'royalblue')

}, fig.width=12, fig.height=6)
```

##### Stage samples

```{r ex2_vd_c3, eval = F}
ae_dsgraveley2011 <- ae(dsgraveley2011$g, r_dsgraveley2011)
ae_dslevin2016dmel <- ae(dslevin2016dmel$g, r_dsgraveley2011)
```

```{r ex2_vd_c3bis, include = F, eval = gen_figs}
# speed up w/ no bootstrap
ae_dsgraveley2011 <- ae(dsgraveley2011$g, r_dsgraveley2011, bootstrap.n = 1)
ae_dslevin2016dmel <- ae(dslevin2016dmel$g, r_dsgraveley2011, bootstrap.n = 1)
```


```{r ex2_vd_c4, echo=F, fig.height=6, fig.width=12}
show_fig(expr = {
  par(mfrow = c(1,2))
  rg <- range(c(ae_dsgraveley2011$age.estimates[,1], dsgraveley2011$p$age))
  plot(ae_dsgraveley2011$age.estimates[,1] ~ dsgraveley2011$p$age, 
       xlab = "Chronological age", ylab = "Estimated age (dsgraveley2011)", 
       xlim = rg, ylim = rg,
       main = "Chron. vs Estimated ages for dsgraveley2011\n(on dsgraveley2011 reference)", lwd = 2)
  points(ae_dsgraveley2011$age.estimates[,1]~dsgraveley2011$p$age, type = 'l', lty = 2)
  
  abline(a = 0, b = 1, lty = 3, lwd = 2)
  legend("bottomright", legend = "x = y", lwd=3, col=1, lty = 3, bty='n')
  
  rg <- range(c(ae_dslevin2016dmel$age.estimates[,1], dslevin2016dmel$p$age))
  plot(ae_dslevin2016dmel$age.estimates[,1] ~ dslevin2016dmel$p$age, 
       xlab = "Chronological age", ylab = "Estimated age (dsgraveley2011)", 
       xlim = rg, ylim = rg,
       main = "Chron. vs Estimated ages for dslevin2016dmel\n(on dsgraveley2011 reference)", lwd = 2)
  # points(ae_dslevin2016dmel$age.estimates[,1] ~ dslevin2016dmel$p$age, type = 'l', lty = 2)
  abline(a = 0, b = 1, lty = 3, lwd = 2)
  
  legend("bottomright", legend = "x = y", lwd=3, col=1, lty = 3, bty='n')
}, fig.height=6, fig.width=12)
```


We notice here that our validation data age estimates vary from chronological age.
However, this is due to the single-embryo nature of the data. 
Indeed, inter-individual variability is shown clearly, where it would otherwise be averaged out in bulk data.
Furthermore, if we look at the dynamics of the `dslevin2016dmel` data, we'll see that chronological age specified for the samples is erroneous, as reflected by noisy expression dynamics.

```{r ex2_vd_c5, eval=gen_figs}
pca_dslevin2016dmel <- stats::prcomp(t(dslevin2016dmel$g), rank = 20,
                                     center = TRUE, scale = FALSE)
```

```{r ex2_vd_c6, echo = F, fig.height=12, fig.width=12}
show_fig(expr = {
  par(mfrow = c(4,4))
  invisible(sapply(c(0,4), function(ipp){
    sapply(seq_len(4), function(i){
      i <- i + ipp
      plot(dslevin2016dmel$p$age, pca_dslevin2016dmel$x[,i], lwd = 2,
           xlab = "age", ylab = "PC", main = paste0("PC", i))
    })
    sapply(seq_len(4), function(i){
      i <- i + ipp
      plot(ae_dslevin2016dmel$age.estimates[,1], pca_dslevin2016dmel$x[,i], lwd = 2, col = "firebrick",
           xlab = "ae", ylab = "PC", main = paste0("PC", i, " (estimated age)"))
      box(col = "firebrick", lwd = 2)
    })
  }))
}, fig.height=12, fig.width=12)
```

This demonstrates a difficulty of producing high-resolution time series due to developmental asynchronicity between samples.