```{r, include = FALSE, eval = F}
knitr::opts_knit$set(out.format = "html")
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%'
)
options(width=100)

gen_figs <- T
figpath <- "../inst/cdoc/RAPToR-refbuilding_figs/"
if(!file.exists(figpath)){
  dir.create(figpath)
}


library(RAPToR)
library(RColorBrewer)
library(beeswarm)

requireNamespace("limma", quietly = T)
requireNamespace("stats")

transp <- function(col, a=.5){
  colr <- col2rgb(col)
  return(rgb(colr[1,], colr[2,], colr[3,], a*255, maxColorValue = 255))
}

png_custom <- function(figname, path = "", 
                       fig.width = 7, fig.height = 5, res = 150, ...){
  png(filename = paste0(path, figname, ".png"), 
      width = fig.width, height = fig.height, res = res, units = "in")
}

show_fig <- function(figname = knitr::opts_current$get("label"), expr, path = figpath, ...){
  if(gen_figs){
    png_custom(figname = figname, path = figpath, ...)
    eval(expr = expr)
    dev.off()
  }
  else{
    knitr::include_graphics(paste0(path, figname, ".png"))
  }
}
```


### The data

We are using two *Danio rerio* (zebrafish) embryonic development time series datasets. 
The dataset used to build the reference has uneven time sampling, as can often be the case.

We show a trick using ranks to build an adequate model in order to avoid interpolation bias. 

The datasets are 

 1. A time series of zebrafish embryonic development (with uneven sampling) published by @white2017high, hereafter called `ds5`. This is the dataset used to build the reference. ([Data accessible in the publication](https://elifesciences.org/articles/30860))
 1. A high-resolution time series of embryonic development published by @levin2016mid, hereafter called `ds6`. This is the dataset used for external validation. (Accession : [GSE60619](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60619))

#### {.tabset}
Code to generate `ds5` and `ds6` :

##### Hide

##### Show
```{r ex3_raw2rpkm_func}
raw2rpkm <- function(X, gene.length, id.col = 1, l.col='length'){
  # Compute RPKM from raw counts
  if(!all(rownames(X)%in%gene.length[, id.col])){
    stop("Some genes are missing length info !")
  }
  res <- sapply(colnames(X), function(samp){
    pm <- sum(X[,samp])/1e6
    rpkm <- (X[,samp]/pm)/(gene.length[match(rownames(X), gene.length[, id.col]), l.col]/1000)
  })
  rownames(res) <- rownames(X)
  return(res)
}
```

```{r load_zeb_genes, eval = F}
requireNamespace("biomaRt", quietly = TRUE)

mart <- biomaRt::useMart("ensembl", dataset = "drerio_gene_ensembl")
zeb_genes <- biomaRt::getBM(attributes = c("ensembl_gene_id", "transcript_length"), 
                               mart = mart)
rm(mart)
```

<br>

###### `ds5`
```{r load_ds5_c, eval = F}
p_url_ds5 <- "http://europepmc.org/articles/PMC5690287/bin/elife-30860-supp1.tsv"

g_url_ds5 <- "http://europepmc.org/articles/PMC5690287/bin/elife-30860-supp2.tsv"
g_file_ds5 <- "../inst/extdata/ds5.tsv"
utils::download.file(g_url_ds5, destfile = g_file_ds5)

X_ds5 <- read.table(g_file_ds5, h = T, sep  ="\t", as.is = T, quote = "\"")
rownames(X_ds5) <- X_ds5$Gene.ID
X_ds5 <- X_ds5[,-(1:8)]


# convert to rpkm & ensembl_id
X_ds5 <- format_ids(X_ds5, zeb_genes, from = "ensembl_gene_id", to = "ensembl_gene_id")
X_ds5 <- raw2rpkm(X = X_ds5, gene.length = zeb_genes, id.col = "ensembl_gene_id", l.col = "transcript_length")


# pheno data
P_ds5 <- read.table(p_url_ds5, h = T, sep = "\t", as.is = T)
P_ds5 <- P_ds5[P_ds5$sequencing == "RNASeq", c("sample", "accession_number", "stage", "stageName", "sampleName")]

# timings of stages from the White et al. eLife (2017) publication of the data.
# time given in hours post-fertilization
timepoints <- data.frame(stage = unique(P_ds5$stageName), 
                         hours_pf = c(0, .75, 2.25, 3, 4.3, 5.25, 6, 8, 10.3, 
                                      16, 19, 24, 30, 36, 48, 72, 96, 120),
                         stringsAsFactors = F, row.names = "stage")
P_ds5$age <- timepoints[P_ds5$stageName, "hours_pf"]
P_ds5$batch <- factor(gsub(".*-(\\d)$", "\\1", P_ds5$sampleName))

X_ds5 <- X_ds5[, P_ds5$sample]

ds5 <- list(g = X_ds5, p = P_ds5)
save(ds5, file = "../inst/extdata/ds5.RData", compress = "xz")

# cleanup
file.remove(g_file_ds5)
rm(p_url_ds5, g_url_ds5, g_file_ds5, X_ds5, P_ds5, timepoints)
```

<br>

###### `ds6`
```{r load_ds6_c, eval = F}
geo_ds6 <- "GSE60619"

g_url_ds6 <- GEOquery::getGEOSuppFiles(geo_ds6, makeDirectory = FALSE, fetch_files = FALSE)
g_file_ds6 <- "../inst/extdata/ds6.txt.gz"
utils::download.file(url = as.character(g_url_ds6$url[2]), destfile = g_file_ds6)

X_ds6 <- read.table(gzfile(g_file_ds6), h = T, sep = '\t', as.is = T, row.names = 1, comment.char = "")

# convert to rpkm & ensembl id
X_ds6 <- format_ids(X_ds6, zeb_genes, from = "ensembl_gene_id", to = "ensembl_gene_id")
X_ds6 <- raw2rpkm(X = X_ds6, gene.length = zeb_genes, id.col = "ensembl_gene_id", l.col = "transcript_length")


# pheno data
P_ds6 <- Biobase::pData(GEOquery::getGEO(geo_ds6, getGPL = F)[[1]])

# filter relevant fields/samples
P_ds6 <- P_ds6[, c("title", "geo_accession", "time (min after fertilization):ch1")]
colnames(P_ds6)[3] <- "time"
P_ds6$title <- as.character(P_ds6$title)

P_ds6 <- P_ds6[P_ds6$title %in% colnames(X_ds6),]
X_ds6 <- X_ds6[, P_ds6$title]

# formatting
P_ds6$title <- gsub('Metazome_ZF_timecourse_', '', P_ds6$title)
colnames(X_ds6) <- P_ds6$title

P_ds6$age <- as.numeric(P_ds6$time) / 60

ds6 <- list(g = X_ds6, p = P_ds6)
save(ds6, file = "../inst/extdata/ds6.RData", compress = "xz")

# cleanup
file.remove(g_file_ds6)
rm(geo_ds6, g_url_ds6, g_file_ds6, X_ds6, P_ds6)
```

```{r ex3_cleanup_genes, eval = F}
rm(zeb_genes, raw2rpkm)
```


#### Normalization & Quick look
```{r ex3_load, include=FALSE, eval=gen_figs}
load("../inst/extdata/ds5.RData")
load("../inst/extdata/ds6.RData")
```

```{r ex3_qnorm_log, eval=gen_figs}
ds5$g <- limma::normalizeBetweenArrays(ds5$g, method = "quantile")
ds5$g <- log(ds5$g + 1)

ds6$g <- limma::normalizeBetweenArrays(ds6$g, method = "quantile")
ds6$g <- log(ds6$g + 1)
```

```{r ex3_ql_c1, results='markup', eval=gen_figs}
ds5$g[1:5, 1:5]
#>                    zmp_ph133_B zmp_ph133_D zmp_ph133_E zmp_ph133_F zmp_ph133_G
#> ENSDARG00000000001    1.992991    1.827910   1.7427071    1.839980   1.7305791
#> ENSDARG00000000002    1.013519    1.049910   0.7850242    1.046494   0.8564017
#> ENSDARG00000000018    2.247738    2.322763   2.0246236    2.159045   2.3384767
#> ENSDARG00000000019    4.198081    4.295515   4.3738184    4.298936   4.3578940
#> ENSDARG00000000068    4.172288    4.226427   4.0332980    4.059632   3.9252796

head(ds5$p, n = 5)
#>        sample accession_number         stage stageName sampleName age batch
#> 1 zmp_ph133_B       ERS1079239 Zygote:1-cell    1-cell   1-cell-1   0     1
#> 2 zmp_ph133_D       ERS1079240 Zygote:1-cell    1-cell   1-cell-2   0     2
#> 3 zmp_ph133_E       ERS1079241 Zygote:1-cell    1-cell   1-cell-3   0     3
#> 4 zmp_ph133_F       ERS1079243 Zygote:1-cell    1-cell   1-cell-4   0     4
#> 5 zmp_ph133_G       ERS1079244 Zygote:1-cell    1-cell   1-cell-5   0     5
```

##### Correlation Matrix
```{r ex3_ql_c2, echo = F, fig.height=5, fig.width=5, out.width="60%"}
show_fig(expr = {
  cor_ds5 <- cor(ds5$g, method = "spearman")
  ord <- order(ds5$p$age)
  heatmap(cor_ds5[ord, ord], Colv = NA, Rowv = NA, scale = "none", keep.dendro = F, margins = c(1,5),
          RowSideColors = transp(as.numeric(ds5$p$batch[ord])), labRow = "", labCol = "")
  par(xpd = T)
  mtext(text = unique(ds5$p$age), side = 1, line = c(3.8, 4), 
        at = seq(-.12, .915, l = length(unique(ds5$p$age))), cex = .6)
  
  # color key
  image(x = c(.95,1), y = seq(0.6,1, l = 10), useRaster = T,
        z = matrix(seq(min(cor_ds5), max(cor_ds5), l = 10), ncol = 10),
        col = hcl.colors(12, "YlOrRd", rev = TRUE), add = T)
  
  text(.975, 1, pos = 3, labels = expression(rho), font = 2)
  text(1, y = seq(0.6,1, l = 10), pos = 4, 
       labels = round(seq(min(cor_ds5), max(cor_ds5), l = 10), 2), cex = .6)
  
  xlp <- 1.025
  batch_legend <- as.character(1:5)
  text(xlp, .5, labels = "batch", font = 2, cex = .8, adj = .5)
  text(xlp, seq(.3,.48, l = 5), labels = batch_legend, adj = 1, pos = 1,
       col = levels(ds5$p$batch), font = 2, cex = .7)
  
  mtext(at = xlp, line = 4, side = 1, text = "(hours)", cex = .8)
}, fig.height=5, fig.width=5)
```

##### Plotting components

```{r ex3_ql_c3, eval = gen_figs}
pca_ds5 <- stats::prcomp(ds5$g, rank = 25)
```

```{r ex3_ql_c4, echo = F, fig.height=6, fig.width=12}
show_fig(expr = {
  par(mfrow = c(2,4))
  invisible(sapply(seq_len(8), function(i){
    plot(ds5$p$age, pca_ds5$rotation[,i], lwd = 2, col = ds5$p$batch,
         xlab = "age", ylab = "PC", main = paste0("PC", i))
    sapply(seq_along(levels(ds5$p$batch)), function(l){
      s <- which(ds5$p$batch == levels(ds5$p$batch)[l])
      points(ds5$p$age[s], pca_ds5$rotation[s,i], col = l, 
             type = 'l', lty = 2)
    })
    # points(ndat$age, pred_ds5_comp[, i], col = "royalblue", type = 'l', lwd = 2)
    if(i == 1)
      legend("bottomright", bty = 'n', legend = batch_legend, title = "Batch",
             pch = c(rep(1, 5)), lty = c(rep(NA, 5)), col = c(1:5), lwd = 3)
  }))
}, fig.height=6, fig.width=12)
```

Notice how the sampling is sparser towards the end of the time series, with dynamics being "wider".
Fitting splines on the components here will lead to a poor fit of the earlier timepoints.

To bypass this issue, we can use ranks instead of the timepoints.

```{r ex3_ql_c5, eval = gen_figs}
# using data.table's rank function to get the "dense" tie method
ds5$p$rank <- data.table::frank(ds5$p$age, ties.method = "dense")
```


```{r ex3_ql_c6, echo = F, fig.height=6, fig.width=12}
show_fig(expr = {
  par(mfrow = c(2,4))
  invisible(sapply(seq_len(8), function(i){
    plot(ds5$p$rank, pca_ds5$rotation[,i], lwd = 2, col = ds5$p$batch,
         xlab = "rank", ylab = "PC", main = paste0("PC", i))
    sapply(seq_along(levels(ds5$p$batch)), function(l){
      s <- which(ds5$p$batch == levels(ds5$p$batch)[l])
      points(ds5$p$rank[s], pca_ds5$rotation[s,i], col = l, 
             type = 'l', lty = 2)
    })
    box(col = "grey50", lwd = 3)
    # points(ndat$age, pred_ds5_comp[, i], col = "royalblue", type = 'l', lwd = 2)
    if(i == 1)
      legend("top", bty = 'n', legend = batch_legend, title = "Batch",
             pch = c(rep(1, 5)), lty = c(rep(NA, 5)), col = c(1:5), lwd = 3)
  }))
}, fig.height=6, fig.width=12)
```

These dynamics will be fitted much more cleanly.
To predict the data in a uniform time scale, we can just pick values on the rank scale such that they translate to a uniform series on the age scale with a simple linear warp, as will be done below.


### Model fitting

#### Component number

```{r ex3_mf_c1, eval=gen_figs}
nc <- sum(summary(pca_ds5)$importance[3,] < .999) + 1
nc
#> [1] 77
```

#### Model

```{r ex3_mf_c2, eval=gen_figs}
m_ds5 <- ge_im(X = ds5$g, p = ds5$p, formula = "X ~ s(rank, bs = 'ds') + batch", nc = nc)
```

```{r ex3_mf_c3, echo = F, eval=gen_figs}
mp_ds5 <- mperf(scale(ds5$g), predict(m_ds5), is.t = T)
as.data.frame(mp_ds5, row.names = "")
```

```{r ex3_mf_print}
#>        aCC       aRE        MSE     aRMSE
#>  0.8772099 0.2281205 0.01266176 0.1125245
```


#### Validation

##### Predict

We'll be using a linear warp to get a uniform time series.

```{r ex3_linwarp, eval=gen_figs}
linwarp <- function(x, xyt, xc = 1, yc = 2){
  # Computes a linear interpolation of input x to y value
  # x = values of x to convert to y
  # xyt = table with known sets of x/y
  # xc, yc = column indices of x and y in xyt
  
  if(min(x) < min(xyt[,xc]) | max(x) > max(xyt[,xc]))
    stop("Some values of x are outside of the known x/y sets")
  
  # set up y to x conversion table
  xyt <- xyt[!duplicated.default(xyt[,xc]),]
  xyt <- xyt[order(xyt[,xc]),]
  
  xyt[,"dify"] <- c(0, diff(xyt[,yc]))
  xyt[,"difx"] <- c(1, diff(xyt[,xc]))
  xyt <- rbind(xyt[1,], xyt) # double 1st line for edge case

  xout <- unlist(sapply(x, function(xi){
    rsup <- which(xyt[-1,xc] >= xi)[1] + 1
    xyt[rsup-1, yc] + (xi - xyt[rsup-1, xc])/xyt[rsup, "difx"] * xyt[rsup, "dify"]
  }))
  
  return(xout)
}
```


```{r ex3_vd_c1, results='markup', eval=gen_figs}
# setup newdat
n.inter <- 200
newdat <- data.frame(age = seq(min(ds5$p[, "age"]), max(ds5$p[, "age"]), l = n.inter),
                     batch = rep("1", n.inter)) # predict as batch 1

# apply linwarp
newdat$rank <- linwarp(newdat$age, xyt = ds5$p[, c("age", "rank")], xc = 1, yc = 2)

head(newdat)
#>         age batch     rank
#> 1 0.0000000     1 1.000000
#> 2 0.6030151     1 1.804020
#> 3 1.2060302     1 2.304020
#> 4 1.8090452     1 2.706030
#> 5 2.4120603     1 3.216080
#> 6 3.0150754     1 4.011596

# predict 
pred_m_ds5 <- predict(m_ds5, newdata = newdat)
pred_m_ds5_comp <- predict(m_ds5, newdata = newdat, as.c = TRUE)
```

We want a uniform series on the age scale, but have to input values on the rank scale in the model which is why we use `linwarp()`.
To give a sense of what the function did, we can plot the ranks against the age. 

```{r ex3_vd_c1.1, echo = F, fig.height=4, fig.width=6, out.width="60%"}
show_fig(expr = {
  plot(rank ~ age, data = ds5$p)
  points(rank ~ age, data = newdat, col = "firebrick", cex = .5, pch = 1, main = "rank vs. age")
  legend("bottomright", legend = c("ds5", "newdat"), bty = "n",
         col = c(1, "firebrick"), pt.cex = c(1,.5), lwd = 2, pch = 1, lty = NA)
}, fig.height=4, fig.width=6)
```


##### Plot component predictions

On the rank scale : 

```{r ex3_vd_c2, echo = F, fig.width=12, fig.height=6}
show_fig(expr = {
par(mfrow = c(2,4))
invisible(sapply(seq_len(8), function(i){
  plot(ds5$p$rank, pca_ds5$rotation[,i], lwd = 2, col = ds5$p$batch,
       xlab = "rank", ylab = "PC", main = paste0("PC", i))
  sapply(seq_along(levels(ds5$p$batch)), function(l){
    s <- which(ds5$p$batch == levels(ds5$p$batch)[l])
    points(ds5$p$rank[s], pca_ds5$rotation[s,i], col = l, 
           type = 'l', lty = 2)
  })
  points(newdat$rank, pred_m_ds5_comp[, i], col = "royalblue", type = 'l', lwd = 3)
  box(col = "grey50", lwd = 3)
  if(i == 1)
    legend("top", bty = 'n', legend = c(batch_legend, "pred"), title = "Batch",
           pch = c(rep(1, 5), NA), lty = c(rep(NA, 5), 1), col = c(1:5, "royalblue"), lwd = 3)
  }))
}, fig.width=12, fig.height=6)
```

Back on the age scale :

```{r ex3_vd_c2bis, echo = F, fig.width=12, fig.height=6}
show_fig(expr = {
  par(mfrow = c(2,4))
  invisible(sapply(seq_len(8), function(i){
    plot(ds5$p$age, pca_ds5$rotation[,i], lwd = 2, col = ds5$p$batch,
         xlab = "age", ylab = "PC", main = paste0("PC", i))
    sapply(seq_along(levels(ds5$p$batch)), function(l){
      s <- which(ds5$p$batch == levels(ds5$p$batch)[l])
      points(ds5$p$age[s], pca_ds5$rotation[s,i], col = l, 
             type = 'l', lty = 2)
    })
    points(newdat$age, pred_m_ds5_comp[, i], col = "royalblue", type = 'l', lwd = 3)
    if(i == 1)
      legend("bottomright", bty = 'n', legend = c(batch_legend, "pred"), title = "Batch",
             pch = c(rep(1, 5), NA), lty = c(rep(NA, 5), 1), col = c(1:5, "royalblue"), lwd = 3)
    }))
}, fig.width=12, fig.height=6)
```




##### Build reference & stage samples

```{r ex3_vd_c3, eval = F}
# make a 'reference object' 
r_ds5 <- list(interpGE = pred_m_ds5, time.series = newdat$age)

ae_ds5 <- ae(ds5$g, r_ds5$interpGE, r_ds5$time.series)
ae_ds6 <- ae(ds6$g, r_ds5$interpGE, r_ds5$time.series)
```

```{r ex3_vd_c3bis, include = F, eval = gen_figs}
# make a 'reference object' 
r_ds5 <- list(interpGE = pred_m_ds5, time.series = newdat$age)

# speed up w/ no bootstrap
ae_ds5 <- ae(ds5$g, r_ds5$interpGE, r_ds5$time.series, bootstrap.n = 1)
ae_ds6 <- ae(ds6$g, r_ds5$interpGE, r_ds5$time.series, bootstrap.n = 1)
```


```{r ex3_vd_c4, echo=F, fig.height=6, fig.width=12}
show_fig(expr = {
  par(mfrow = c(1,2))
  rg <- range(c(ae_ds5$age.estimates[,1], ds5$p$age))
  plot(ae_ds5$age.estimates[,1]~ds5$p$age, 
       xlab = "Chronological age", ylab = "Estimated age (ds5)", 
       xlim = rg, ylim = rg,
       main = "Chron. vs Estimated ages for ds5 (on ds5 reference)", lwd = 2, col = factor(ds5$p$batch))
  invisible(sapply(levels(factor(ds5$p$batch)), function(l){
    s <- ds5$p$batch == l
    points(ae_ds5$age.estimates[s,1]~ds5$p$age[s], type = 'l', 
           lty = 2, col = which(l==levels(factor(ds5$p$batch))))
  }))
  
  abline(a = 0, b = 1, lty = 3, lwd = 2)
  legend("bottomright", legend = c(batch_legend, "x = y"), 
         lwd=3, col=c(1:5, 1), bty='n', pch = c(rep(1,5),NA), lty = c(rep(NA, 5), 3))
  
  # rg <- range(c(ae_ds6$age.estimates[,1], ds6$p$age))
  plot(ae_ds6$age.estimates[,1]~ds6$p$age, 
       xlab = "Chronological age", ylab = "Estimated age (ds5)", 
       xlim = rg, ylim = rg,
       main = "Chron. vs Estimated ages for ds6 (on ds5 reference)", lwd = 2)
  # points(ae_ds6$age.estimates[,1] ~ ds6$p$age, type = 'l', lty = 2)
  abline(a = 0, b = 1, lty = 3, lwd = 2)
  
  legend("bottomright", legend = "x = y", lwd=3, col=1, lty = 3, bty='n')
}, fig.width=12, fig.height=6)
```


##### On a model without using ranks

We'll build the same model (not the optimal model !) without considering uneven sampling, for comparison.

```{r ex3_vd_c5, eval=gen_figs}
m_ds5_norank <- ge_im(X = ds5$g, p = ds5$p, formula = "X ~ s(age, bs = 'ds') + batch", nc = nc)
```

```{r ex3_vd_c5bis, echo=F, eval=gen_figs}
mp_ds5_norank <- mperf(scale(ds5$g), predict(m_ds5_norank), is.t = T)
as.data.frame(mp_ds5_norank, row.names = "")
```

```{r ex3_vd_mf_print}
#>        aCC       aRE        MSE     aRMSE
#>  0.8477562 0.2153231 0.01803523 0.1342953
```


```{r ex3_vd_c6, eval=gen_figs}
pred_m_ds5_norank <- predict(m_ds5_norank, newdata = newdat)
pred_m_ds5_comp_norank <- predict(m_ds5_norank, newdata = newdat, as.c = TRUE)
```

Let's plot the components on the rank and age scales, as before.

```{r ex3_vd_c7, echo = F, fig.width=12, fig.height=6}
show_fig(expr = {
  par(mfrow = c(2,4))
  invisible(sapply(seq_len(8), function(i){
    plot(ds5$p$rank, pca_ds5$rotation[,i], lwd = 2, col = ds5$p$batch,
         xlab = "rank", ylab = "PC", main = paste0("PC", i))
    sapply(seq_along(levels(ds5$p$batch)), function(l){
      s <- which(ds5$p$batch == levels(ds5$p$batch)[l])
      points(ds5$p$rank[s], pca_ds5$rotation[s,i], col = l, 
             type = 'l', lty = 2)
    })
    points(newdat$rank, pred_m_ds5_comp_norank[, i], col = "firebrick", type = 'l', lwd = 3)
    box(col = "grey50", lwd = 3)
    if(i == 1)
      legend("top", bty = 'n', legend = c(batch_legend, "pred (no rk)"), title = "Batch",
             pch = c(rep(1, 5), NA), lty = c(rep(NA, 5), 1), col = c(1:5, "firebrick"), lwd = 3)
    }))
}, fig.width=12, fig.height=6)
```

Back on the age scale :

```{r ex3_vd_c7bis, echo = F, fig.width=12, fig.height=6}
show_fig(expr = {
  par(mfrow = c(2,4))
  invisible(sapply(seq_len(8), function(i){
    plot(ds5$p$age, pca_ds5$rotation[,i], lwd = 2, col = ds5$p$batch,
         xlab = "age", ylab = "PC", main = paste0("PC", i))
    sapply(seq_along(levels(ds5$p$batch)), function(l){
      s <- which(ds5$p$batch == levels(ds5$p$batch)[l])
      points(ds5$p$age[s], pca_ds5$rotation[s,i], col = l, 
             type = 'l', lty = 2)
    })
    points(newdat$age, pred_m_ds5_comp_norank[, i], col = "firebrick", type = 'l', lwd = 3)
    if(i == 1)
      legend("bottomright", bty = 'n', legend = c(batch_legend, "pred (no rk)"), title = "Batch",
             pch = c(rep(1, 5), NA), lty = c(rep(NA, 5), 1), col = c(1:5, "firebrick"), lwd = 3)
    }))
}, fig.width=12, fig.height=6)
```

We can already see that the model has trouble predicting the dynamics accurately.
For example, we pick up noise in PC6 and flatten dynamics in PC7.

This has consequences on the estimates *of the validation dataset*, as you'll see when we stage the samples.

```{r ex3_vd_c8, eval = F}
# make a 'reference object' 
r_ds5_norank <- list(interpGE = pred_m_ds5_norank, time.series = newdat$age)

ae_ds5_norank <- ae(ds5$g, r_ds5_norank$interpGE, r_ds5_norank$time.series)
ae_ds6_norank <- ae(ds6$g, r_ds5_norank$interpGE, r_ds5_norank$time.series)
```

```{r ex3_vd_c8bis, include = F, eval=gen_figs}
# make a 'reference object' 
r_ds5_norank <- list(interpGE = pred_m_ds5_norank, time.series = newdat$age)

ae_ds5_norank <- ae(ds5$g, r_ds5_norank$interpGE, r_ds5_norank$time.series, bootstrap.n = 1)
ae_ds6_norank <- ae(ds6$g, r_ds5_norank$interpGE, r_ds5_norank$time.series, bootstrap.n = 1)
```


```{r ex3_vd_c9, echo=F, fig.height=6, fig.width=13}
show_fig(expr = {
  par(mfrow = c(1,2))
  rg <- range(c(ae_ds5$age.estimates[,1], ds5$p$age))
  plot(ae_ds5_norank$age.estimates[,1]~ds5$p$age, 
       xlab = "Chronological age", ylab = "Estimated age (ds5)", 
       xlim = rg, ylim = rg,
       main = "Chron. vs Estimated ages for ds5 (on ds5 norank reference)", lwd = 2, col = factor(ds5$p$batch))
  invisible(sapply(levels(factor(ds5$p$batch)), function(l){
    s <- ds5$p$batch == l
    points(ae_ds5_norank$age.estimates[s,1]~ds5$p$age[s], type = 'l', 
           lty = 2, col = which(l==levels(factor(ds5$p$batch))))
  }))
  
  abline(a = 0, b = 1, lty = 3, lwd = 2)
  legend("bottomright", legend = c(batch_legend, "x = y"), 
         lwd=3, col=c(1:5, 1), bty='n', pch = c(rep(1,5),NA), lty = c(rep(NA, 5), 3))
  
  # rg <- range(c(ae_ds6$age.estimates[,1], ds6$p$age))
  plot(ae_ds6_norank$age.estimates[,1]~ds6$p$age, 
       xlab = "Chronological age", ylab = "Estimated age (ds5)", 
       xlim = rg, ylim = rg,
       main = "Chron. vs Estimated ages for ds6 (on ds5 norank reference)", lwd = 2)
  # points(ae_ds6$age.estimates[,1] ~ ds6$p$age, type = 'l', lty = 2)
  abline(a = 0, b = 1, lty = 3, lwd = 2)
  
  legend("bottomright", legend = "x = y", lwd=3, col=1, lty = 3, bty='n')
}, fig.height=6, fig.width=13)
```

The "gaps" or "steps" you can see on the validation dataset's estimates are due to interpolation bias : the picked up noise and flattened dynamics mentioned above.
Essentially, the model errors create local "unlikely/unrealistic" gene expression zones, which will not correlate well with the samples of corresponding age.
These zones will most often find themselves in the gaps between timepoints of the dataset used to build the reference, meaning the estimates of its samples are unaffected.

While we've used a suboptimal model here (which clearly displays the model errors on component plots), some interpolation bias can be much more subtle. 
In such cases, this is something that can only be assessed using an external dataset. 