---
title: "`RAPToR` - Real Age Prediction from Transcriptome staging On Reference"
output: 
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    code_folding: hide
  rmarkdown::pdf_document:
    toc: true
    number_sections: yes
    includes:
      in_header: "../inst/tex/preamble.tex"
subtitle: "`r BiocStyle::pkg_ver('RAPToR')`"
vignette: >
  %\VignetteIndexEntry{RAPToR}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
bibliography: ../inst/REFERENCES.bib
author: Romain Bulteau
date: "`r Sys.setlocale('LC_TIME', 'C'); format(Sys.Date(), '%B %Y')`"
---
  
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%',
  cache = T
)
options(width=70)

figpath <- "../inst/cdoc/RAPToR_figs/"

library(ggplot2)
library(ggpubr)
requireNamespace("limma", quietly = T)
```


<br>
<br>

`RAPToR` is a computational framework to estimate the real age of biological samples from gene expression. This is especially useful for fast-developing organisms -- such as *C. elegans* worms, flies, or zebrafish -- where many factors substantially impact developmental speed, and thus unintended developmental variation between samples could obscure or confound the effect of variables of interest.

With `RAPToR` and the inferred age of your samples, you can

 - precisely estimate the effect of perturbations on developmental timing (including in a tissue-specific way), 
 - increase power in differential expression analyses, 
 - quantify differential expression due to uncontrolled development and,
 - recover perturbation-specific effects on gene expression even when completely confounded by development. 


<br>

Please cite our article (@bulteau2022real) if you use RAPToR in your research:

 - Bulteau, R., Francesconi, M. Real age prediction from the transcriptome with RAPToR. *Nat Methods* (2022). (https://doi.org/10.1038/s41592-022-01540-0)


# Quickstart

Given an expression matrix `X` of transcripts per million (TPM) with samples as columns and genes as rows.
```{r qs_load_data, include=FALSE}
load("../inst/extdata/dshendriks2014.RData")
X <- dshendriks2014$g[, 1:6]
```

```{r qs_1, class.source="fold-show", message=FALSE, warning=FALSE}
library(RAPToR)
library(wormRef) # reference data package

# quick look at expression matrix
X[1:5, 1:3] 
```

Load an appropriate reference.
```{r qs_ref, class.source="fold-show"}
ref <- prepare_refdata("Cel_larv_YA", "wormRef", 600)
```

Stage samples.
```{r qs_stage, class.source="fold-show"}
ae_X <- ae(X, ref)
```

Show output.
```{r qs_print, class.source="fold-show"}
print(ae_X)
```
Plot age estimates.
```{r qs_plot, class.source="fold-show", fig.width=6, fig.height=3, out.width="80%"}
plot(ae_X)
```
Extract age estimates.
```{r qs_get, class.source="fold-show", fig.width=6, fig.height=3, out.width="50%"}
ae_X$age.estimates[,1]
```

# About `RAPToR`
## Why use `RAPToR` ?

In gene expression data, unknown and unintended developmental variation among biological samples can obscure and confound the effect of variables of interest. 
As many factors influence growth speed (including experimental conditions), synchronizing samples can be challenging but failing to do has a strong impact on gene expression.

Aware of this problem, studies with large scale developmental profiling generally re-order or rank the samples post-profiling with methods such as BLIND (@anavy2014blind) that combine dimension-reduction (PCA, Diffusion Map, ...) and a trajectory-finding method.
Unfortunately, this only works with hundreds of samples and/or time-series designs and thus doesn't apply to the overwhelming majority of expression profiling studies. 

Overcoming this limitation, `RAPToR` provides a way to precisely determine the real age of *single samples* from their expression profile.


## How does it work ?

`RAPToR` is a 2-step process:

 1. A reference gene expression time-series is interpolated to build a near-continuous, **high-temporal-resolution reference** (a number of which are included in associated data-packages, see below).
 1. A **correlation profile** against this reference is dressed for of each of your samples, and the timing of the correlation peak is the estimated age.  **Bootstrapping on genes** then gives a confidence interval of the estimates.


```{r fig_tool_overview, echo=F, fig.wide=TRUE}
knitr::include_graphics(file.path(figpath,"tool_overview.jpg"), auto_pdf = T)
```

## What type of data can be used ?

`RAPToR` can stage samples from any genome-wide gene expression profiling method  : RNA-seq (preferably as TPM), MicroArray, ...

Note that the references provided in the data-packages are $log(X+1)$ of expression values, so applying this transformation to your data is important if comparing expression changes with the reference (but is not required for staging).  

***Data must not be gene-centered***, as this destroys the relationship between gene levels within a sample.


## General structure of the package

The main package (`RAPToR`) holds the functions needed for staging samples and building references.

To provide a quick and easy way to infer the age of samples, we pre-built several references for commonly used organisms from available data in the literature.
References are voluminous for an R package, so they are stored in separate "data-packages". 
For example, `wormRef` stores the *C. elegans* references.

References available in a data-package are listed with the `list_refs()` function (you must have the data-package installed for this, 
[`wormRef` can be installed from here.](https://github.com/LBMC/wormRef))

```{r list_refs, class.source="fold-show"}
list_refs("wormRef")
```

The data-packages currently available are listed in the [README of RAPToR's github repo](https://github.com/LBMC/RAPToR).
You can also build your own reference data-packages following guidelines given in the [vignette on this topic](RAPToR-datapkgs.html).


# Detailed usage example

In this part, we show how to stage two *C. elegans* time-series datasets using references from `wormRef` (which you will need installed to use its references, [see here for installation](https://github.com/LBMC/RAPToR)).

 1. A larval development time-series of 4 different strains published by @aeschimann2017lin41, hereafter called `dsaeschimann2017`.
 1. A high-resolution time-series of late larval development published by @hendriks2014extensive, hereafter called `dshendriks2014`

Both datasets are available on GEO 
([GSE80157](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE80157), 
[GSE52861](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52861)).


Code to create the `dsaeschimann2017` and `dshendriks2014` objects can be found [at the end of this vignette](#load_dsaeschimann2017_dshendriks2014).


```{r load_ds, include=FALSE, eval=gen_figs}
load("../inst/extdata/dsaeschimann2017.RData")
load("../inst/extdata/dshendriks2014.RData")
```



Here is a quick look at the data 
```{r head_dsaeschimann2017, class.source="fold-show",}
dsaeschimann2017$g[1:5,1:3]

head(dsaeschimann2017$p, n = 5)
```

```{r head_dshendriks2014, class.source="fold-show"}
dshendriks2014$g[1:5,1:3]

head(dshendriks2014$p, n = 5)
```

We quantile-normalize and log the expression data (this is not required for staging, but is useful for plotting the expression data).
```{r quantile_norm_log, class.source="fold-show"}
dsaeschimann2017$g <- limma::normalizeBetweenArrays(dsaeschimann2017$g, method = "quantile")
dsaeschimann2017$g <- log1p(dsaeschimann2017$g) # log1p(x) = log(x + 1)

dshendriks2014$g <- limma::normalizeBetweenArrays(dshendriks2014$g, method = "quantile")
dshendriks2014$g <- log1p(dshendriks2014$g)
```


You may need to convert probe IDs or gene IDs of your data to match those of the reference series.
For example, all the references included in the `wormRef` data-package use WormBase Gene IDs (*e.g.* `WBGene00016153`).

Transcript-level data is aggregated into gene-level expression in our references for a broader usage (since it is possible to compute gene-level expression from transcripts, but not the other way around).
RNASeq expression data should ideally be sum-aggregated at the count level, and other types of data -- such as microarray, or RNASeq TPM (if counts are unavailable) -- are mean-aggregated.

To help with this conversion, we provide the `format_ids()` function and gene ID reference tables are included in the data-packages (built directly from [`biomaRt`](https://www.ensembl.org/index.html) queries).



## Choosing a reference dataset

`RAPToR` estimates sample age using correlation with a reference time-series, meaning you must select the proper reference to compare your samples with. 
You can determine which reference of a data-package covers appropriate development for your samples from the list with `list_refs()` like above, or by using the `plot_refs()` function.

If we haven't built references for your favorite organism yet, you can take a look at [the *Building your own references* section](#building-your-own-references) in this vignette for a quick-start guide or [the dedicated reference-building vignette](RAPToR-refbuilding.html) for a more in-depth explanation.

The chart below shows the references available in the `wormRef` data-package, along with landmark developmental stages. 

```{r plot_refs, fig.height=6, fig.width=9, class.source="fold-show"}
plot_refs("wormRef")
```



## Loading the reference

To increase the accuracy of age estimates beyond the temporal resolution of time-series profiling experiments, we interpolate on gene expression dynamics of the reference with respect to time.
Pre-built references with optimal interpolation parameters can be loaded using `prepare_refdata()`.

The `Cel_larval` reference would be appropriate for both example datasets since their samples range from mid-larval to early young adult.
Interpolation resolution is given through either `n.inter` (number of interpolated points) or `by.inter` (interpolation step in reference time units). 
In the interest of lightening the computational load (each interpolated timepoint is compared to the samples) you can choose smaller `n.inter` or larger `by.inter` values. We find having `n.inter` > 500 gets the most out of interpolation.


```{r prep_ref, class.source="fold-show"}
r_larv <- prepare_refdata("Cel_larval", "wormRef", n.inter = 600)
```

Note that age estimates will be given in the time unit and scale of the chosen reference (here, hours post-hatching at 20$^\circ$C).

```{r show_ref, class.source="fold-show"}
r_larv
```



## Age estimation

All that is left to do is run the `ae()` (**a**ge **e**stimation) function.

```{r include=FALSE}
set.seed(1) # for reproducibility
```


```{r ae_dsaeschimann2017, class.source="fold-show", results='hold'}
ae_dsaeschimann2017 <- ae(dsaeschimann2017$g,      # input gene expression matrix
                          r_larv)                  # reference object
```

```{r ae_dshendriks2014, class.source="fold-show", results='hold'}
ae_dshendriks2014 <- ae(samp = dshendriks2014$g,      # input gene expression matrix
                        r_larv)                       # reference object
```

Let's look at the results.


```{r plot_ae1, class.source="fold-show", fig.height=9, fig.width=8, out.width="80%"}
plot(ae_dsaeschimann2017, groups = dsaeschimann2017$p$strain, 
     lmar = 14, g.line = 3)
```


```{r plot_ae2, class.source="fold-show", fig.height=4, fig.width=8, out.width="80%"}
plot(ae_dshendriks2014)
```
At 25$^\circ$C, *C. elegans* worms develop 1.5 times faster than worms at 20$^\circ$C.
Since the worms we staged were grown at 25Â°C, but the reference -- and thus, the inferred age -- corresponds to 20$^\circ$C development,
we can see this 1.5 factor by fitting a simple linear model between chronological and estimated age.


```{r lm_ds, class.source="fold-show"}
lm_dsaeschimann2017 <- lm(ae_dsaeschimann2017$age.estimates[,1] ~ dsaeschimann2017$p$age)
summary(lm_dsaeschimann2017)$coefficients

lm_dshendriks2014 <- lm(ae_dshendriks2014$age.estimates[,1] ~ dshendriks2014$p$age)
summary(lm_dshendriks2014)$coefficients
```

```{r plot_compae, fig.height=6, fig.width=12, echo=F}
par(mfrow = c(1,2), pty='s', mar=c(5,5,4,1))

# plot first dataset
plot(ae_dsaeschimann2017$age.estimates[,1]~dsaeschimann2017$p$age, 
     xlab = "Chronological age (25C)", ylab = "RAPToR age estimates (20C)", 
     main = "Chronological vs estimated age\nfor dsaeschimann2017", lwd = 2, 
     col = factor(dsaeschimann2017$p$strain))
invisible(sapply(levels(factor(dsaeschimann2017$p$strain)), function(l){
  s <- dsaeschimann2017$p$strain == l
  points(ae_dsaeschimann2017$age.estimates[s,1]~dsaeschimann2017$p$age[s], 
         type = 'l', 
         lty = 2, col = which(l==levels(factor(dsaeschimann2017$p$strain))))
}))
abline(lm_dsaeschimann2017, lwd=2, col=2)
legend("bottomright", 
       legend = c("let-7", "lin-41", "let-7/lin-41", 
                  "N2", "lm_dsaeschimann2017"), 
       lwd=3, col=c(1:4, 2), bty='n', 
       pch = c(1,1,1,1,NA), lty = c(rep(NA, 4), 1))

# plot second dataset
plot(ae_dshendriks2014$age.estimates[,1]~dshendriks2014$p$age, 
     xlab = "Chronological age (25C)", ylab = "RAPToR age estimates (20C)", 
     main = "Chronological vs estimated age\nfor dshendriks2014", lwd = 2)
points(ae_dshendriks2014$age.estimates[,1]~dshendriks2014$p$age, 
       type = 'l', lty = 2)
abline(lm_dshendriks2014, lwd=2, col=2)

legend("bottomright", legend = "lm_dshendriks2014", lwd=3, col=2, bty='n')
```


## Understanding the output

The output of `ae()` is an `ae` object containing (among other things) the age estimate and confidence intervals from booststrapping (age estimates on random gene subsets).

<br>
General information can be accessed via `summary()`.
```{r summ_ae, class.source="fold-show",}
summary(ae_dshendriks2014)
```

<br>
Age estimates and their confidence intervals are accessible through `$age.estimates`, a table with

 - `age.estimate`, the global estimate for the sample (whole gene set).
 - `lb`, `ub`, the lower and upper bounds of the bootstrapped age estimates' confidence interval (Median Absolute Deviation).
 - `cor.score`, the correlation score between the sample and reference at the age estimate.

```{r show_ae, class.source="fold-show",}
head(ae_dshendriks2014$age.estimates)
```



Estimates and confidence intervals can be displayed in the form of a 'dotchart' with the default `plot()` function (as done above).

The `ae` object also holds record of correlation scores of each samples with the full reference span (for bootstrap estimates as well). 
These correlation profiles can be plotted with `plot_cor()`

```{r plot_cor_print, class.source="fold-show", fig.height=5, fig.width=8, out.width="80%"}
par(mfrow=c(2,2))
plot_cor(ae_dshendriks2014, subset = c(1,4,9,14))
```

Red bars correspond to the estimate confidence interval, and the 95% interval of bootstrap correlation with the reference is shown as black dotted lines. 
The sample age estimate is displayed below the interval.





# Building your own references

This section gives a broad overview, on reference-building, but note [there is a vignette entirely dedicated to it.](RAPToR-refbuilding.html).

Interpolating on time-series expression data is the key to get the high-temporal-resolution references we need for `RAPToR`.
When using our pre-built references, interpolation is done internally by the `prepare_refdata()` function, using the functions described below. 

To build a reference, you will require a time-series of gene expression data for your favorite organism.
We will use the `dsaeschimann2017` data from the previous examples to illustrate the process.

## The gene expression interpolation model

Gene expression interpolation models (GEIMs) are built with the `ge_im()` function.
This function takes as input 3 key arguments : 

 - `X` : your time-series gene expression matrix (genes as rows, samples as columns)
 - `p` : a dataframe of phenotypic data, samples as rows *in the same order as `X` columns*. This should include the age/time variable and any other covariate(s) you want to include in the model (*e.g* batch, strain)
 - `formula` : the model formula. This should be a standard R formula using terms found in `p`, which may include elements (such as splines) from chosen model type (see below). **It must start with `X ~`**.

For example, using `dsaeschimann2017`, we build the following model:
```{r model_dsaeschimann2017, class.source="fold-show"}
m_dsaeschimann2017 <- ge_im(X = dsaeschimann2017$g, 
                            p = dsaeschimann2017$p, 
                            formula = "X ~ s(age, bs = 'ts') + strain", 
                            nc = 32)
```



In order to model a large number of output variable (genes), our strategy is to project the data in a dimensionally-reduced space and interpolate there before re-projecting the data back to genes.
We do this with Principal Components or Independant Components ([Independant Component Analysis](https://en.wikipedia.org/wiki/Independent_component_analysis) ).

Both PCA and ICA perform the same type of linear transformation on the data (they just optimize different criteria). We get the following :

$$
X_{(m\times n)} = G_{(m\times c)}S^{T}_{(n\times c)}
$$
with $X$, the matrix of $m$ genes by $n$ samples, $G$ the gene loadings ($m$ genes by $c$ components) and $S^T$ the sample scores ($n$ samples by $c$ components). 
When performing PCA (or ICA) on gene expression data, $S$ is what's usually plotted (e.g. PC1 vs. PC2) to see how samples are grouped in the component space. 

@alter2000singular demonstrated that singular value decomposition of gene expression data can be taken as "eigengenes", giving a global picture of the expression landscape and dynamics with a few components. 
We use this property through GEIMs.
We fit a model on the columns of $S^T$ (eigengenes), predict in the component space, and reconstruct the gene expression data by a matrix product with the gene loadings. 

We've implemented 2 model types : Generalized Additive Models (GAMs, the default) and Generalized Linear Models (GLMs). 
GAMs rely on the `gam()` function of the [`mgcv`](https://cran.r-project.org/web/packages/mgcv/index.html) package, and GLMs on the `glm()` function of the `stats` core package. 
The specified model formula can make use of all the tools one can use with `gam()` or `glm()`, most notably the variety of polynomial or smoothing splines implemented through the `s()` function for GAMs.

If you are unfamiliar with the `mgcv` package or GLMs in R, we recommend you look at their function documentations (especially `s()` for GAMs) to understand what terms can be included in the formulas to model non-linear dynamics.

Note that a single model formula is specified and applied to all the components, but models are fitted independently on the components.


## Defining the appropriate model formula and parameters

### Model type

Our default GEIM fits GAMs on PCA components, which is a robust choice when applying a smoothing spline to the data.
PCA and ICA interpolation usually yield near-identical results (but ICA components tend to be more biologically meaningful or interpretable).

### Parameter estimation

The number of components to use for the interpolation is by default set to the number of samples.
However, we recommend setting a cutoff on explained variance of PCA components. 
For example, on the `dsaeschimann2017` dataset, we set a threshold at $99\%$ :

```{r pca_dsaeschimann2017, class.source="fold-show"}
pca_dsaeschimann2017 <- stats::prcomp(t(dsaeschimann2017$g), 
                                      center = TRUE, scale = FALSE, 
                                      rank = 25)
nc <- sum(summary(pca_dsaeschimann2017)$importance[3,] < .99) + 1
nc
```

This threshold should be set with respect to the noise in the data. For example, in very noisy data, would you consider that $99\%$ of the variance in the dataset corresponds to meaningful information or dynamics ?
One can also keep only components that have '*intelligible dynamics*' with respect to time, defined as those where a model fit explains $> 0.5$ of the deviance (noisy components with no dynamics have poor fits).

<br>

Choosing from different splines (and/or parameters) can be done with Cross-Validation (CV) through the `ge_imCV()` function, which inputs `X`, `p` and a `formula_list` to test. 
Other parameters on the CV itself can also be given (*e.g.* training set size).

Below is an example to choose among available spline types for the `dsaeschimann2017` GEIM.
```{r ss2, include = F}
set.seed(2)
```

```{r cv_dsaeschimann2017, class.source="fold-show"}
smooth_methods <- c("tp", "ts", "cr", "ps")
flist <- as.list(paste0("X ~ s(age, bs = \'", smooth_methods, "\') + strain"))

cat(unlist(flist), sep='\n') # print formulas to test

cv_dsaeschimann2017 <- ge_imCV(X = dsaeschimann2017$g, 
                               p = dsaeschimann2017$p, 
                               formula_list = flist,
                               cv.n = 20, nc = nc, 
                               nb.cores = 3)
```

Plot the results.
```{r plot_cv_dsaeschimann2017_print, fig.height=6, fig.width=9, out.width="80%"}
plot(cv_dsaeschimann2017, names = paste0("bs = ", smooth_methods), 
     outline = F, swarmargs = list(cex = .8), boxwex=.5)
```

`ge_imCV()` computes multiple indices of model performance :
the average Correlation Coefficient (`aCC`), the average Relative Error (`aRE`), Mean Squared Error (`MSE`) and average Root MSE (`aRMSE`). 
These indices all compare model predictions with the true data, and they are computed both using the validation set (CV Error, cve) *and* on the training set (Model PerFormance, mpf).

From the plots above, we can see the different splines perform similarly and all could work for the reference. 
We choose `ts` (a thin-plate regression spline), as it minimizes CV error without much impact model performance. 


## Building a reference object from a model

A `ref` object is built from a GEIM using `make_ref()`, specifying interpolation resolution and relevant metadata.

On our `dsaeschimann2017` example :
```{r pred_dsaeschimann2017, class.source="fold-show"}
r_dsaeschimann2017 <- make_ref(
  m = m_dsaeschimann2017,
  n.inter = 100,                  # interpolation resolution
  t.unit = "h past egg-laying",   # time unit
  cov.levels = list("strain" = "N2"), # covariate lvls to use for interpolation
  metadata = list("organism" = "C. elegans", # any metadata
                  "profiling" = "whole-organism, bulk",
                  "technology" = "RNAseq")
  ) 

```


## Checking/Validating the interpolation

After building a reference, we can check interpolation results by: 

 - Checking model fits on components (plots)
 - Staging the samples on their interpolated reference, or better (if possible) stage another independent time-series on your reference for external validation.

We can do both with our example, using the `dshendriks2014` dataset for external validation.

Plotting the model with the reference object (or equivalent metadata) shows component interpolation, with deviance explained (DE) and relative error (RE) for each component. The fit values are also returned by the plot function.

```{r plot_pca_dsaeschimann2017, class.source="fold-show", warning=FALSE, message=FALSE, fig.width=10, fig.height=5, out.width="80%"}
par(mfrow = c(2,4))
fit_vals <- plot(m_dsaeschimann2017, r_dsaeschimann2017, ncs=1:8, 
                 col = dsaeschimann2017$p$strain, col.i = 'royalblue')

head(fit_vals)
```


Of note, we are predicting model values as `N2` (lightblue). While all strains are shown on the plots, some model parameters depend on the selected `N2` strain. 

```{r ae_dsaeschimann2017_test, class.source="fold-show", message=FALSE, warning=FALSE, results='hide'}
ae_test_dsaeschimann2017 <- ae(dsaeschimann2017$g, r_dsaeschimann2017)
ae_test_dshendriks2014 <- ae(dshendriks2014$g, r_dsaeschimann2017)
```


```{r plot_comp_aeref, echo=F, fig.height=6, fig.width=12}
par(mfrow = c(1,2), pty='s', mar=c(5,5,4,1))

rg <- range(c(ae_test_dsaeschimann2017$age.estimates[,1], dsaeschimann2017$p$age))
plot(ae_test_dsaeschimann2017$age.estimates[,1]~dsaeschimann2017$p$age,
     xlab = "Chronological age", ylab = "RAPToR age estimates (r_aeschimann2017)",
     xlim = rg, ylim = rg,
     main = "Chron. vs Estimated ages for dsaeschimann2017\n(on dsaeschimann2017 reference)", lwd = 2, col = factor(dsaeschimann2017$p$strain))
invisible(sapply(levels(factor(dsaeschimann2017$p$strain)), function(l){
  s <- dsaeschimann2017$p$strain == l
  points(ae_test_dsaeschimann2017$age.estimates[s,1]~dsaeschimann2017$p$age[s], type = 'l',
         lty = 2, col = which(l==levels(factor(dsaeschimann2017$p$strain))))
}))

abline(a = 0, b = 1, lty = 3, lwd = 2)
legend("bottomright", legend = c("let-7", "lin-41", "let-7/lin-41", "N2", "x = y"),
       lwd=3, col=c(1:4, 1), bty='n', pch = c(1,1,1,1,NA), lty = c(rep(NA, 4), 3))

rg <- range(c(ae_test_dshendriks2014$age.estimates[,1], dshendriks2014$p$age))
plot(ae_test_dshendriks2014$age.estimates[,1]~dshendriks2014$p$age,
     xlab = "Chronological age", ylab = "RAPToR age estimates (r_aeschimann2017)",
     xlim = rg, ylim = rg,
     main = "Chron. vs Estimated ages for dshendriks2014\n(on dsaeschimann2017 reference)", lwd = 2)
points(ae_test_dshendriks2014$age.estimates[,1] ~ dshendriks2014$p$age, type = 'l', lty = 2)
abline(a = 0, b = 1, lty = 3, lwd = 2)

legend("bottomright", legend = "x = y", lwd=3, col=1, lty = 3, bty='n')
```


<br>


# Using a prior

When few genes are available for staging (e.g. for [tissue-specific staging](RAPToR-showcase.html#capturing-tissue-specific-development)), it may be appropriate to give RAPToR a prior to help age estimation.
In the `ae()` function, priors work by giving the parameters for gaussian distributions of time (for each sample). The correlation peaks are then ranked according to the prior's density.
The correlation profile is unaffected by the prior, only the choice of the correlation peak is.

**This implies that with a prior which is completely off, the estimate may also be wrong ; use with care.**

Priors are given *in the time scale of the reference*, so beware of growth speed difference with temperature or different time origins (fertilization, egg-laying, hatching...). 
For example, the `dshendriks2014` *C. elegans* data used above is grown at 25$^\circ$C, resulting in a different growth speed than the reference which is at 20$^\circ$C development. 

We insist on careful use due to these possible differences and bias introduced by the prior. 

A first run without priors will give a general idea of the difference between the chronological and developmental age of your samples.

Once the priors are determined, you will need to set the standard deviation of the gaussian centered on the sample with the `prior.params` argument. 
This parameter will also indirectly change the weight of the prior over the correlation score for estimate selection.

On our `dshendriks2014` example, we can use adjusted known chronological ages for 20$^\circ$C.
```{r hend_prior, class.source="fold-show"}
# rough approximation based on our previous lm
priors <- dshendriks2014$p$age * 1.6 - 5 

ae_dshendriks2014_prior <- ae(samp = dshendriks2014$g, 
                              r_larv,
                              prior = priors,
                              prior.params = 10)
```


```{r plot_ae_prior_print, class.source="fold-show", fig.height=4, fig.width=8, out.width="80%"}
plot(ae_dshendriks2014_prior, 
     main="Age estimates with priors on dshendriks2014",
     show.prior = T, col.p = 'red', l.pos = 'bottomright')

# check prior estimates are equal to previous ones
all(ae_dshendriks2014_prior$age.estimates[,1]==ae_dshendriks2014$age.estimates[,1])
```

As expected, here the estimates are identical to those without priors. 


# Code to generate `dsaeschimann2017` and `dshendriks2014`

**Note : set the `data_folder` variable to an existing path on your system where you want to store the objects.**

```{r load_namespaces, eval = F}
data_folder <- "../inst/extdata/"

requireNamespace("wormRef", quietly = T)
requireNamespace("utils", quietly = T)
requireNamespace("GEOquery", quietly = T) # May need to be installed with bioconductor
requireNamespace("Biobase", quietly = T) # same
```

**Utility functions**

```{r c2tpm_func, code = readLines("convert2tpm.R"), echo = TRUE, eval=FALSE}
```

**dsaeschimann2017**

```{r load_dsaeschimann2017_c, code = readLines("load_dsaeschimann2017.R"), echo = TRUE, eval=FALSE}
```

**dshendriks2014**

```{r load_dshendriks2014_c, code = readLines("load_dshendriks2014.R"), echo = TRUE, eval=FALSE}
```



<hr>
# References {.unnumbered}
<div id="refs"></div>

<hr>
# SessionInfo {.unnumbered}
```{r sessInfo, collapse=FALSE, class.source = 'fold-show'}
sessionInfo()
```
