---
title: "`RAPToR` - Data-packages"
output: 
  rmarkdown::html_document :
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{RAPToR-datapkgs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: ../inst/REFERENCES.bib
author: Romain Bulteau
date: "`r format(Sys.Date(), '%B %Y')`"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%'
)
options(width=100)

library(RAPToR)
```

# Who is this vignette for ?

This vignette is aimed at those who've already familiarized themselves with [`RAPToR` reference building](RAPToR-refbuilding.html).
It's also for us to keep track of how to continue improving `RAPToR` with new references.

If you've already built a few references and want to make them available to the world (or use them more easily yourself), you're in the right place.

You will need some basic knowledge of R package development. Data-packages are, after all, *packages*. 
This document only details how to set up your data-package to interact properly with `RAPToR`.


# What's a data-package ?

By definition, a "data-package" is an R package in which one stores large datasets (over a few Mo, at least).
This is a good practice for several reasons. 

 1. If your data rarely or never changes, updates to the data-package (and thus, download of the data) will be minimal. If included in a standard package, large data can be a burden during install.
 1. The data may never be used. Why have users download data they won't need ?
 1. CRAN standards limit package size to 5MB (documentation included). A large dataset is better off when separated from methods and functions that may need it.
 
Hadley Wickam gives thorough advice on organizing data in packages in his [*R packages* book](http://r-pkgs.had.co.nz/data.html).


# RAPToR data-packages

`RAPToR` relies on references to operate. 
These can sometimes be tedious to build, so we want to give users access to pre-built references.
A reference can be boiled down to :

 - gene expression data, 
 - optimal interpolation parameters, and
 - some metadata.

To avoid overcharging the end user, we split references to data-packages.
The data is then used through the `RAPToR` interface described below.

## Reference data
Reference data are stored as `.RData` objects, which should
include everything needed to construct the interpolated reference (as listed above).
There can be as many as you like in one data-package. We keep one organism per data-package.


Being consistent, clear and concise with naming can help users (or yourself !) find their way around references.
For example, `wormRef` references are named with an organism code `Cel` (*C. elegans*) followed by the developmental period covered by the reference *e.g.* `larval`.


The structure of the `Cel_larval` reference object is detailed below as an example.

A list with

 * `$ g` The gene expression matrix (genes as rows, samples as columns).
 * `$ p` A dataframe of phenotypic data on the samples :
   * `sname` sample names,
   * `age`   *developmental* age of the samples (scaled),
   * `cov`   covariate, factor indicating which of 3 time series,
   * `age_ini`  *chronological* age of the samples,
   * `accession` sample accession ID for GEO.
 * `$ geim_params`  A list with necessary parameters for interpolation :
   * `$ formula = "X ~ s(age, bs = 'ds') + cov"`
   * `$ method = "gam"`
   * `$ dim_red = "pca"`
   * `$ nc = 40`
 * `$ t.unit = "h past egg-laying (20C)"` the time unit.
 * `$ cov.levels` A named list with covariate levels to interpolate as.
   *`$ cov = "O.20"`
 * `$ metadata` A named list with any extra metadata
   * `$ organism = "C. elegans"`
   * `$ profiling = "whole-organism, bulk"`
   * `$ technology = "RNAseq"`



**Document the references.** It is standard practice to document data.
*What is the data? Where is the data from?* etc.

## Data-package interface with RAPToR

To have users access data-package information and references directly from `RAPToR`, we've set up a standard system using reference names.

A few objects are **necessary** for this interface to work.


### `.prepref_` functions

`.prepref_` functions (note the "dot") are the key of the interface : they **prep**are the **ref**erence for the user.
They must respect the naming convention `.prepref_ref_name()` (*e.g.* `.prepref_Cel_larval()`) *and* take `n.inter` and/or `by.inter` as arguments.

These functions are the backbone called by `prepare_refdata()` when fetching a reference, and thus should
output the reference `ref` object with the specified parameters.
This means building the GEIM model, and calling `make_ref()` with the appropriate parameters and metadata. 

We have made a [function factory](https://adv-r.hadley.nz/function-factories.html) to generate these functions. 
It inputs the reference data object described above, and returns the corresponding prepref function. 
```{r prepref, eval = F}
.prepref_skel <- function(data, from=NULL, to=NULL){
  # .prepref function factory
  f <- function(n.inter=NULL, by.inter=NULL){
    m <- RAPToR::ge_im(
      X = data$g,
      p = data$p,
      formula = data$geim_params$formula,
      method = data$geim_params$method,
      dim_red = data$geim_params$dim_red,
      nc = data$geim_params$nc
    )
    return(RAPToR::make_ref(m, 
                            n.inter = n.inter,
                            by.inter = by.inter,
                            from = from, 
                            to = to,
                            t.unit = data$t.unit,
                            cov.levels = data$cov.levels,
                            metadata = data$metadata)
    )
  }
  return(f)
}
```

To make `.prepref_Cel_larval()`, we simply include the following code in the data-package, along with the function factory.

```{r prepref_larv, eval=F}
.prepref_Cel_larval <- .prepref_skel(wormRef::Cel_larval)
```


### `ref_list` object

`RAPToR` expects a `ref_list` object in the data-package. 
This is what's displayed when calling the `list_refs(datapkg)` function.

```{r list_ref, results='markup'}
list_refs(datapkg = "wormRef")
```

The form/layout of this object is free, but reference names should be included somewhere, as the user needs them to access the reference through `prepare_refdata()`.



### `.plot_refs()` function

This function is optional, but very useful to guide users to the correct reference for their samples.
Including a `.plot_refs()` function (note the “dot”) in a data-package will allow it to be called by the `plot_refs(datapkg)` function in `RAPToR`.

```{r plot_ref, fig.width = 8, fig.height=6, out.width="75%"}
plot_refs(datapkg = "wormRef")
```

### Other objects

You're free to include any extra objects in your data-package that may be useful.
For example, the `wormRef` package has a `Cel_devstages` object with information on key developmental stages of *C. elegans*, which is used for building the plot in `.plot_refs()` (*cf.* above).


<br>
<br>
<br>