```{r, include = FALSE, eval = T}
knitr::opts_knit$set(out.format = "html", header = "")
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%'
)
options(width=100)

gen_figs <- F
figpath <- "../inst/cdoc/RAPToR-showcase_figs/"
if(!file.exists(figpath)){
  dir.create(figpath)
}


library(RAPToR)
library(RColorBrewer)
library(beeswarm)
library(parallel)
library(vioplot)

library(limma)
requireNamespace("wormRef")
library(stats)

transp <- function(col, a=.5){
  colr <- col2rgb(col)
  return(rgb(colr[1,], colr[2,], colr[3,], a*255, maxColorValue = 255))
}

png_custom <- function(figname, path = "", 
                       fig.width = 7, fig.height = 5, res = 150, ...){
  png(filename = paste0(path, figname, ".png"), 
      width = fig.width, height = fig.height, res = res, units = "in")
}

show_fig <- function(figname = knitr::opts_current$get("label"), expr, path = figpath, ...){
  if(gen_figs){
    png_custom(figname = figname, path = figpath, ...)
    eval(expr = expr)
    dev.off()
  }
  else{
    knitr::include_graphics(paste0(path, figname, ".png"))
  }
}
```

## The data

The dataset used in this example corresponds to a *C. elegans* profiling experiment by @lehrbach2012post, hereafter called `dslehrbach2012` (Accession : [E-MTAB-1333](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-1333/)). 
Profiling was done using microarray (Affymetrix). 


The experimental setup is the following. 
Two *C. elegans* strains : wild-type (`wt`) and pash-1(mj100) (`mut`) were profiled at 0, 6, 12 and 24 hours past the L4 stage (~ 50 hours post-hatching). These were done in triplicates.

In this scenario, performing a Differential Gene Expression (DGE) analysis to uncover the effect of the strain relies on the inclusion of the effect of development.
This is true in any time series context, even more so here because the late larval development of *C. elegans* is known to lead to drastic changes in gene expression within a short time frame.

#### {.tabset}

<a name="code-gen_ds1-ds2"></a>
  
Code to generate `dslehrbach2012` :
  
##### Hide
  
##### Show

**Note : set the `data_folder` variable to an existing path on your system where you want to store the objects.**

```{r sc1_load_namespaces, eval = F}
data_folder <- "../inst/extdata/"

library("biomaRt") # May need to be installed with bioconductor
library("limma")   # ..
library("affy")    # ..
library("gcrma")   # ..

requireNamespace("wormRef", quietly = T)
requireNamespace("utils", quietly = T)
```

###### `dslehrbach202`
```{r load_dslehrbach_c, eval = F}
# get probe ids from the arrayexpress
probe_ids <- read.table("https://www.ebi.ac.uk/arrayexpress/files/A-AFFY-60/A-AFFY-60.adf.txt", h=T,
                        comment.char = "", sep = "\t", skip = 17, as.is = T)


# pheno data
p_url <- "https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-1333/E-MTAB-1333.sdrf.txt"
p <- read.table(p_url, h = T, sep = "\t", comment.char = "", as.is = T)

# geno data
g_url <- unique(p$Comment..ArrayExpress.FTP.file.)
g_dir <- paste0(data_folder, "dslehrbach2012_raw/")
g_file <- paste0(g_dir, "raw.zip")

if(!file.exists(g_file)){
  dir.create(g_dir)
  utils::download.file(g_url, destfile = g_file)
}
utils::unzip(g_file, exdir = g_dir)

g <- affy::ReadAffy(filenames = p$Array.Data.File, celfile.path = g_dir, phenoData = p)
g <- affy::expresso(g, bg.correct = F, normalize = F,
                    pmcorrect.method = "pmonly", summary.method = "median")
g <- 2^exprs(g) # expresso log2s the data
colnames(g) <- p$Source.Name

# format ids
g <- RAPToR::format_ids(g, probe_ids, from = 1, to = 7)
g <- RAPToR::format_ids(g, wormRef::Cel_genes, from = 3, to = 1)


# filter relevant fields
p <- p[, c(1,22,24)]
colnames(p) <- c("title", "tpastL4", "strain")
p$strain <- factor(p$strain, levels = c('wild type', 'pash-1(mj100)'), labels = c('wt', 'strain'))
p$rep <- factor(gsub("\\w+_h\\d+\\.(\\d)","\\1", p$title))

dslehrbach2012 <- list(g = g, p = p)
save(dslehrbach2012, file = paste0(data_folder, "dslerhbach2012.RData"), compress = "xz")

# cleanup
file.remove(g_file)
unlink(g_dir, recursive = T)

rm(g, g_url, g_dir, g_file, p, p_url, probe_ids)
```

<br>


```{r sc1_load, include=F, eval=gen_figs}
load("../inst/extdata/dslerhbach2012.RData")
```

## Normalization

```{r sc1_qnormlog, eval = gen_figs}
dslehrbach2012$g <- limma::normalizeBetweenArrays(dslehrbach2012$g, method = "quantile")
dslehrbach2012$g <- log(dslehrbach2012$g + 1)
```


## Staging the samples

The samples' (chronological) ages range from after L4 to 24h past L4.
The `Cel_YA_2` reference from `wormRef` covers this range, so we'll select it to stage the samples.

```{r sc1_ae_lb, eval=gen_figs}
r_ya <- prepare_refdata("Cel_YA_2", datapkg = "wormRef", n.inter = 200)

ae_dslb <- ae(dslehrbach2012$g, r_ya$interpGE, r_ya$time.series)
```



```{r sc1_plot_ae, fig.width=8, fig.height=6,  echo = F}
show_fig(expr = { 
  plot(ae_dslb, groups = as.factor(dslehrbach2012$p$tpastL4), subset = rev(1:24), 
     main = "age estimates of samples grouped by timepoint",
     show.boot_estimates = T)
}, fig.width=8, fig.height=6)
```

We can see quite a bit of variation in the sample timings, especially at the 0 and 6 hour timepoints.
Another interesting effect can be seen if we group the samples by replicate.

```{r sc1_plot_ae2, fig.width=8, fig.height=6,  echo = F}
show_fig(expr = {
plot(ae_dslb, groups = dslehrbach2012$p$rep, subset = rev(1:24), 
     main = "age estimates of samples grouped by replicate",
     show.boot_estimates = T)
}, fig.width=8, fig.height=6)
```


Notice how, in the first two replicates, the `mut` samples are consistantly older than the `wt` ones, which is the opposite way around for the 3rd replicate.
We can maybe speculate about a potential effect of their sampling order.


## The impact of development on gene expression

```{r sc1_cor_dev, eval = gen_figs}
cc <- cor(dslehrbach2012$g, dslehrbach2012$g, method = "spearman")
```

Let's look at the correlation between the samples and compare it to their estimated age difference.
In the graphic below, we plot the correlation score between all sample pairs against their estimated age difference. 
Each point corresponds to a pair of samples and the color-coding corresponds to whether both samples are of the same strain or not.

```{r sc1_cor_dev_plot, fig.height=5, fig.width=5, echo = F, out.width="60%"}
show_fig(expr = {
  ut <- upper.tri(cc, diag = F)
  ya <- cc[ut]
  xa <- ae_dslb$age.estimates[col(cc)[ut],1] - ae_dslb$age.estimates[row(cc)[ut],1]
  cols <- (dslehrbach2012$p$strain[col(cc)[ut]] != dslehrbach2012$p$strain[row(cc)[ut]]) + 1
  
  plot(xa, ya, ylab = "Spearman correlation",
       xlab = "Estimated age difference",
       cex=.5, xlim = range(xa),
       lwd=2, col = cols)
  # abline(h=0, col='grey50', lty=2)
  legend("bottomleft", legend = c("same strain", "different strain"), 
         title = "Sample pair of", bty = 'n', 
         col = 1:2, pch = 1, lwd = 3, lty = NA, text.font = 2, text.col = 1:2)
}, fig.height=5, fig.width=5)
```

Notice how the strain does not really impact correlation between samples, compared to developmental differences.
We can also observe this through correlation heatmaps ordered by strain & chronological age or estimated age.

<br>

<div style = "display:block">
<div style="width:49%; float:left">
```{r sc1_hm_strain, echo = F, fig.height=5, fig.width=6}
show_fig(expr = {
  heatmap(cc, Colv = NA, Rowv = NA, scale = "none", keep.dendro = F, margins = c(2,5),
          RowSideColors = transp(as.numeric(dslehrbach2012$p$strain)), labRow = "", labCol = "",
          main = "Ordered by strain and chronological timepoint")
  par(xpd = T)
  mtext(c(unique(dslehrbach2012$p$tpastL4), ""), side = 1, line = 4, cex = .8, at = seq(0.1,.95, l = 5))
  mtext("timepoint", side = 1, at = -.1, line = 4, cex = .8)
}, fig.height = 5, fig.width = 6)
```

</div>
<div style="width:49%; float:right">
```{r sc1_hm_age, echo = F, fig.height=5, fig.width=6}
show_fig(expr = {
  ord2 <- order(ae_dslb$age.estimates[,1])
  heatmap(cc[ord2, ord2], Colv = NA, Rowv = NA, scale = "none", keep.dendro = F, margins = c(2,5),
          RowSideColors = transp(as.numeric(dslehrbach2012$p$strain[ord2])), labRow = "", labCol = "",
          main = "Ordered by age estimates")
par(xpd = T)

# color key
image(x = c(.9,.95), y = seq(0.6,1, l = 10), useRaster = T,
      z = matrix(seq(min(cc), max(cc), l = 10), ncol = 10),
      col = hcl.colors(12, "YlOrRd", rev = TRUE), add = T)
text(.925, 1.01, pos = 3, labels = expression(rho), font = 2, cex = 1.25)
text(.95, y = seq(0.6,1, l = 10), pos = 4,
     labels = round(seq(min(cc), max(cc), l = 10), 2), cex = .6)

# strain legend
mtext("Strain", side = 4, line = -1.25, las = 2, at = .45, adj = .5)
mtext(text = c("wt", "mut"), side = 4, line = -1.25, at = c(.4, .45)-.06,
      font = 2, col = transp(1:2, a = .8), las = 2, adj = .5, cex = 1.25)
}, fig.height = 5, fig.width = 6)
```
</div>
</div>
<div style="clear: both;"></div>


<br>
<br>

## Differential Gene Expression analysis

Let's perform a simple DGE analysis using limma, with the following model : 

$$X \sim strain \times \texttt{ns(age, df = 2)} $$

We're using a spline with $\texttt{ns()}$ to tackle non-linear dynamics in the data.

We'll look at the number of genes differentially expressed due to development or strain, which translates to comparing the following nested models (2 vs. 1 for development, 3 vs. 1 for strain).

$$
\begin{align}
Y & = \beta_0 +
\beta_1 I_{strain} +
(\alpha_1 age_{sp1} + \alpha_2 age_{sp2}) +
(\gamma_1 I_{strain} age_{sp1} + \gamma_2 I_{strain} age_{sp2})\\
Y & = \beta_0 + \beta_1 I_{strain}\\
Y & = \beta_0 + (\alpha_1 age_{sp1} + \alpha_2 age_{sp2})
\end{align}
$$

Note that we're not interested in whether genes are up or down-regulated, but only in the detection of an effect. 
Genes will be considered differentially expressed with Bonferoni-Holm adjusted p.values of coefficients below 0.05.

#### {.tabset}
DGE code :

##### Hide
##### Show
```{r sc1_dge_setup_funcs, eval = gen_figs}
# Predictions from limma model
pred_lmFit <- function(Fit){
  tcrossprod(Fit$coefficients, Fit$design)
}

# Compute Goodness of Fit
GoF <- function(Fit, X){
  pred <- pred_lmFit(Fit)
  res <- (X - pred)
  ss <- apply(X, 1, function(ro) sum((ro - mean(ro))^2))
  Rsq <- sapply(seq_len(nrow(X)), function(i){
    1 - sum(res[i,]^2)/ss[i]
  })
  return(Rsq)
}
```

```{r sc1_dge_main_func, eval = gen_figs}
DGE <- function(X, strain, age, df = 2, name = NULL, return.model = FALSE){
  require(splines)
  if(! length(strain) == ncol(X) | ! length(age) == ncol(X))
    stop("strain and age must be of length ncol(X).")
  
  # make pdat df
  pdat <- data.frame(strain = factor(strain), age = as.numeric(age), row.names = colnames(X))
  
  # build design matrix
  d <- model.matrix(~ 1 + ns(age, df = df) * strain, data = pdat) 
  
  # fix colnames
  colnames(d) <- c("b0", paste0(rep("a", df), 1L:df), "strainmut", paste0(rep("strainmut.a", df), 1L:df))
  
  # build contrast matrices for mut and age tests
  if(df == 2){
    cm.mut <- makeContrasts(mut = strainmut, #strainwt - strainmut,
                            mut.i1 = strainmut.a1,
                            mut.i2 = strainmut.a2,
                            levels = d)
    cm.age <- makeContrasts(a1, a2,
                            a.i1 = strainmut.a1,
                            a.i2 = strainmut.a2,
                            levels = d)
  }
  
  if(df == 3){
    cm.mut <- makeContrasts(mut = strainmut, #strainwt - strainmut,
                            mut.i1 = strainmut.a1,
                            mut.i2 = strainmut.a2,
                            mut.i3 = strainmut.a3,
                            levels = d)
    cm.age <- makeContrasts(a1, a2, a3,
                            a.i1 = strainmut.a1,
                            a.i2 = strainmut.a2,
                            a.i3 = strainmut.a3,
                            levels = d)
  }
  
  # fit model
  m.0 <- lmFit(object = X, design = d)
  # get GoF
  gof <- GoF(Fit = m.0, X = X)
  
  # find DE genes for mut
  m.m <- contrasts.fit(m.0, contrasts = cm.mut)
  m.m <- eBayes(m.m)
  
  # find DE genes for age
  m.a <- contrasts.fit(m.0, contrasts = cm.age)
  m.a <- eBayes(m.a)
  
  Tmut <- topTable(m.m, adjust.method = "BH", number = Inf, sort.by = "none")[, c("F", "P.Value", "adj.P.Val")]
  Tage <- topTable(m.a, adjust.method = "BH", number = Inf, sort.by = "none")[, c("F", "P.Value", "adj.P.Val")]
  
  res <- list(gof  = gof,
              tmut = Tmut,
              tage = Tage,
              name = name)
  if(return.model)
    res$model = m.0
  
  rm(m.0, m.m, m.a, Tmut, Tage, d, X, pdat, gof)
  gc(verbose = F)
  return(res)
}
```

<br>
<br>

#### {}

We want to perform *the same DGE analysis* using the chronological age, and our estimates, to see if there is an improvement in detecting strain and age effect.

```{r do_dge, eval = gen_figs, results='hide'}
# format gdata
X <- log2(exp(dslehrbach2012$g))

dge.ca <- DGE(X = X, strain = dslehrbach2012$p$strain, age = dslehrbach2012$p$tpastL4, 
              name = "dge.ca", return.model = T)
dge.ae <- DGE(X = X, strain = dslehrbach2012$p$strain, age = ae_dslb$age.estimates[,1], 
              name = "dge.ae", return.model = T)
```

```{r sc1_plot_dge_func, echo = F}
plot_DGE_pval_vs <- function(pv.x, pv.y, 
                             xlab = "-log10(pv.x)", ylab = "-log10(pv.y)", 
                             main = "DE genes with pv.x vs pv.y", ...)
{ 
  cols <- rowSums(cbind((pv.x < 0.05), (pv.y < 0.05),
                     (pv.x > 0.05)+1 & (pv.y < 0.05)))+1
  
  plot(-log10(pv.x), -log10(pv.y), cex=.3, pch=16, 
       xlab = xlab, ylab = ylab, main = main,
       col = cols, ...)
  
  abline(v=-log10(0.05), h=-log10(0.05), col=2)
  
  legend('bottomright', legend = paste(table(cols)), text.col = 1:4,
         text.font = 2, horiz = T, bty='n')
  invisible(cols)
}

```

In the plots below are shown the adjusted p.values for detection of an effect for each gene, when using chronological or estimate age in the model. The red bars correspond to the 0.05 threshold for significance.

```{r sc1_plot_dge, echo = F, fig.height=5, fig.width=10}
show_fig(expr = { 

par(mfrow = c(1,2))
cols <- plot_DGE_pval_vs(dge.ca$tmut$adj.P.Val, dge.ae$tmut$adj.P.Val, main = "Strain effect detection (chron. vs ae)", 
                         xlab = "-log10(pval) with chron. age", ylab = "-log10(pval) with ae")
# plot(dge.ca$gof, dge.ae$gof, pch = 16, cex = .4, col = cols)
# points(dge.ca$gof[cols>1], dge.ae$gof[cols>1], pch = 16, cex = .6, col = cols[cols>1])

cols <- plot_DGE_pval_vs(dge.ca$tage$adj.P.Val, dge.ae$tage$adj.P.Val, main = "Development effect detection (chron. vs ae)",
                         xlab = "-log10(pval) with chron. age", ylab = "-log10(pval) with ae")
# plot(dge.ca$gof, dge.ae$gof, pch = 16, cex = .4, col = cols)
# points(dge.ca$gof[cols>1], dge.ae$gof[cols>1], pch = 16, cex = .6, col = cols[cols>1])
}, fig.height=5, fig.width=10)
```

```{r dge_increase, eval = gen_figs, results='markup'}
mut_dif.ca <- sum(dge.ca$tmut$adj.P.Val < 0.05)
mut_dif.ae <- sum(dge.ae$tmut$adj.P.Val < 0.05)

age_dif.ca <- sum(dge.ca$tage$adj.P.Val < 0.05)
age_dif.ae <- sum(dge.ae$tage$adj.P.Val < 0.05)

100 * (mut_dif.ae - mut_dif.ca) / (mut_dif.ca) # mut pct increase 
#> [1] 91.8583

100 * (age_dif.ae - age_dif.ca) / (age_dif.ca) # age pct increase
#> [1] 9.081006

```

We detect nearly *twice* as many DE genes for strain using our estimates compared to using the chronological age, and around $10 \%$ more genes with development.
This shows how crucial it can be to take developmental dynamics into account correctly in DGE analyses.


```{r dge_write, echo = F, eval = F, include = F}
write.csv(data.frame(" " = rownames(dge.ca$tmut)),
          file = "../inst/extdata/full_gl.txt", 
          row.names = F, quote = F)


write.csv(data.frame(" " = rownames(dge.ca$tmut)[dge.ca$tmut$adj.P.Val<0.05]),
          file = "../inst/extdata/ca_mut_gl.txt", 
          row.names = F, quote = F)

write.csv(data.frame(" " = rownames(dge.ae$tmut)[dge.ae$tmut$adj.P.Val<0.05]),
          file = "../inst/extdata/ae_mut_gl.txt",
          row.names = F, quote = F)
```


We can also see this increase in the model's fit itself, by looking at the Goodness of Fit distribution accross genes.
The GoF computed is an $R^2 = 1 - \frac{SS_{res}}{SS_{tot}}$ per gene.

```{r sc1_gof_dist, echo = F, fig.height=4, fig.width=6, out.width="60%"}
show_fig(expr = {
  d.ca <- density(dge.ca$gof, from = 0, to = 1)
  d.ae <- density(dge.ae$gof, from = 0, to = 1)
  
  plot(d.ca$x, d.ca$y, type = "l", lwd = 3, ylim = range(c(d.ca$y, d.ae$y)), 
       xlab = "GoF", col = "red", ylab = "Density", main = paste("Model Goodness of Fit distribution across",nrow(X),"genes"))
  points(d.ae$x, d.ae$y, type = 'l', col = "blue", lwd = 3) 
  legend("bottom", lwd = 4, col = c("red", "blue"), 
         legend = c("Model with chron. age", "Model with est. age"),
         text.font = 2, text.col = c("red", "blue"), bty = 'n')
}, fig.height = 4, fig.width = 6)
```


## Comparing model fit between chronological age, estimated age, and random noise on chronological age


If the asynchronicity observed with age estimates is erroneous, then using random noise around the chronological values should yield similar results to using our estimates in the model.
To test this, we can simulate age sets with random noise of similar distribution to the age differences observed between chronological and estimated age.

To compare the results, we use the model's Goodness of Fit (GoF) per gene. As done above, the GoF computed is an $R^2 = 1 - \frac{SS_{res}}{SS_{tot}}$.

We also look at the number of differentially expressed genes found for strain and development (BH-adjusted p.value below 0.05).

```{r sc1_comp_dge1, eval = gen_figs}
age_diffs <- (dslehrbach2012$p$tpastL4 + 50) - ae_dslb$age.estimates[,1]
# Note : + 50 to shift tpastL4 to age, avoiding generation of negative tpastl4 values,
#        this has no impact on the DGE analysis (it just shifts the age window).

# estimate density function of age_diffs
d_ad <- density(age_diffs)

# generate age sets of with random age_diffs-like noise 
set.seed(10)

n <- 100
rd_ages <- lapply(seq_len(n), function(i){
  (dslehrbach2012$p$tpastL4 + 50) + sample(x = d_ad$x, size = nrow(dslehrbach2012$p), prob = d_ad$y, replace = T) 
})

```


```{r sc1_comp_dge1bis, eval = F, include = F}
# plot(density(unlist(lapply(rd_ages, function(a) a - 50- dslehrbach2012$p$tpastL4))))
hist(unlist(lapply(rd_ages, function(a) a - 50 - dslehrbach2012$p$tpastL4)),
     border = transp(1, 0.4), breaks = 50, freq = F, 
     main = "Simulated age differences", xlab = "Age difference")
points(d_ad, type = 'l', lwd = 3)
legend("topright", legend = "Density of age_diffs", lwd = 3, bty = "n")
# hist(age_diffs, col = transp(4), freq = F, add = T,
#      breaks = 20)
# plot(dslehrbach2012$p$tpastL4, ae_dslb$age.estimates[,1], type = "n", ylim = range(rd_ages))
# points(dslehrbach2012$p$tpastL4, ae_dslb$age.estimates[,1], col = 4, lwd = 2)
```


```{r sc1_comp_dge2, eval = gen_figs, results='hide'}
# setup cluster for parallelization
cl <- makeCluster(3, "FORK")

# do DGE on all age sets
rd_dges <- parLapply(cl, seq_len(n), function(i){
  cat("\r", i,"/",n)
  DGE(X, dslehrbach2012$p$strain, rd_ages[[i]], name = paste0("rd.",i))
})

stopCluster(cl)
gc()
```


```{r sc1_format_forplot, echo = F, eval=gen_figs}
qts <- seq(0,1, length.out = 100)
quants <- lapply(seq_len(n), function(i){
  quantile(rd_dges[[i]]$gof, probs = qts)
})
quants <- do.call(rbind, quants)
```

```{r sc1_plot_resdge, echo = F, fig.height=6, fig.width=10}
show_fig(expr = {

layout(mat = matrix(c(2,3,1,1), ncol = 2), widths = c(.45,.55))

plot(0:1, 0:1, type = "n", ylab = "GoF >= quantile", xlab = "quantile", main = "QQ-plot of Goodness of Fit distribution")
invisible(apply(quants, 1, function(ro) points(qts, ro, type = 'l', col = transp("black", a = .4))))
points(qts, quantile(dge.ca$gof, probs = qts), type = "l", lwd = 4, col = 2)
points(qts, quantile(dge.ae$gof, probs = qts), type = "l", lwd = 4, col = 4)

legend('bottomright', legend = c("Chronological age", "Estimated age", "Chron.Age + noise"), 
       col = c(2,4, 1), lwd = c(4,4,1), bty = "n")


diffgens <- lapply(seq_len(n), function(i){
  data.frame(mut = sum(rd_dges[[i]]$tmut$adj.P.Val < 0.05), age = sum(rd_dges[[i]]$tage$adj.P.Val < 0.05))
})

diffgens <- do.call(rbind, diffgens)

hist(diffgens$mut, breaks = 50, main = "Nb of DE genes for strain", xlim = c(0,3500), xlab = "Nb. genes")
abline(v = sum(dge.ca$tmut$adj.P.Val < 0.05), lwd = 3, col = 2)
abline(v = sum(dge.ae$tmut$adj.P.Val < 0.05), lwd = 3, col = 4)

hist(diffgens$age, breaks = 50, xlim = c(min(diffgens$age), 12000), main = "Nb of DE genes for devpt.", xlab = "Nb. genes")
abline(v = sum(dge.ca$tage$adj.P.Val < 0.05), lwd = 3, col = 2)
abline(v = sum(dge.ae$tage$adj.P.Val < 0.05), lwd = 3, col = 4)
}, fig.height=6, fig.width=10)
```

In the plots above, we can see that using the estimated age leads to better model fits and an increased number of differentially expressed genes.

We can also look at the GoF distributions ordered by chronological age quantiles.
The gene GoF are binned along the quantile scale, and shown as boxplots instead of scatterplot for a clearer view. 

The Chronological age GoF QQ line is the same as for the previous plot (since the GoF values are ordered according to it).

```{r sc1_plot_resdge2, echo = F, eval = gen_figs}
ord <- order(dge.ca$gof)
# par(mfrow = c(1,1))
# plot(dge.ca$gof[ord], type = 'l')


gofs <- lapply(seq_len(n), function(i){
  rd_dges[[i]]$gof[ord]
})
gofs <- do.call(rbind, gofs)
```

```{r sc1_plot_resdge3, echo = F, eval = gen_figs}
plot_QQGoFbox <- function(gof, bins = 10, at = seq(0,1,l = bins), boxwex = .5/bins, add = F, plotargs = list(), ...){
  grps <- cut(1:length(gof), bins)
  
  bxs <- lapply(levels(grps), function(l){
    s <- which(grps == l)
    return(boxplot(gof[s], plot = F))
  })
  if(!add){
    do.call(plot, c(list(x = range(at), y=range(gof), type = "n"), plotargs))
  }
  
  invisible(lapply(seq_len(bins), function(i){
    bxp(z = bxs[[i]], at = at[i], add = T, frame.plot = F, axes = F, boxwex = boxwex, ...)
  }))
}
```

```{r sc1_plot_resdge4, echo = F, fig.height=8, fig.width=16}
show_fig(expr = {
  
par(mfrow = c(1,2), mar = c(4,4,4,1))
plot(0:1, 0:1, type = "n", ylab = "GoF", xlab = "quantile", main = "Goodness of Fit distribution in chron. age quantile order\n(estimated age)")
points(qts, quantile(dge.ca$gof, probs = qts), type = "l", lwd = 4, col = 2)
plot_QQGoFbox(dge.ae$gof[ord], bins = 50, add = T, cex = .2, lwd = 1, boxcol=4, medcol = 4, medlwd=6, boxwex = .8/50)

legend('bottomright', legend = c("Chronological age GoF quantiles"), 
       col = c(2), lwd = c(4), bty = "n")


plot(0:1, 0:1, type = "n", ylab = "GoF >= quantile", xlab = "quantile", main = paste0("Goodness of Fit distribution in chron. age quantile order\n(joined simulated age + noise sets, n = ",n, ")"))
points(qts, quantile(dge.ca$gof, probs = qts), type = "l", lwd = 4, col = 2)
plot_QQGoFbox(c(gofs), bins = 50, add = T, cex = .2, lwd = 1, boxwex = .8/50)

}, fig.height=8, fig.width=16)
```


We can see that using our estimates overall outperforms the random noise sets *and* the chronological age.  