```{r, include = FALSE, eval = T}
knitr::opts_knit$set(out.format = "html", header = "")
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%'
)
options(width=100)

gen_figs <- F
figpath <- "../inst/cdoc/RAPToR-showcase_figs/"
if(!file.exists(figpath)){
  dir.create(figpath)
}


library(RAPToR)
library(RColorBrewer)
library(beeswarm)
library(parallel)
# library(vioplot)

library(limma)
requireNamespace("wormRef")
library(stats)

transp <- function(col, a=.5){
  colr <- col2rgb(col)
  return(rgb(colr[1,], colr[2,], colr[3,], a*255, maxColorValue = 255))
}

png_custom <- function(figname, path = "", 
                       fig.width = 7, fig.height = 5, res = 150, ...){
  png(filename = paste0(path, figname, ".png"), 
      width = fig.width, height = fig.height, res = res, units = "in")
}

show_fig <- function(figname = knitr::opts_current$get("label"), expr, path = figpath, ...){
  if(gen_figs){
    png_custom(figname = figname, path = figpath, ...)
    eval(expr = expr)
    dev.off()
  }
  else{
    knitr::include_graphics(paste0(path, figname, ".png"))
  }
}
```


Sometimes, profiled samples may not be from a well-studied organism, with an abundance of reference time series available.
However, you can still refer to the closest model organisms as a reference.
Though results will not be on par as estimates using a reference from the same species, you can stil get good results thanks to the conserved nature of developmental processes across species.

Let's stage samples cross-species using orthologs genes.

### The data

We'll be working with 3 datasets here.

 - 2 embryonic timecourse datasets published by @levin2016mid, for *D. melanogaster* and  *C. elegans*, hereafter called `dslevin2016dmel` and `dslevin2016cel` respectively. (Accessions : [GSE60471](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60471) and [GSE60755](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60755))
 - A Drosophila embryonic development time series, part of the modENCODE project and published by @graveley2011developmental, hereafter called `dsgraveley2011`. (Data downloaded from [fruitfly.org](https://fruitfly.org/sequence/download.html))

Furthermore, we'll use a set of orthologs between *D. melanogaster* and  *C. elegans*, from the supplementary data of @li2014comparison. This list will be stored in the `glist` object.

Since it can get a little confusing between the plots below, the 
<span style="color:firebrick"> ***D.melanogaster*** **samples will always be in red** </span>
and the <span style="color:darkblue"> ***C. elegans*** **samples will always be in blue** </span>

### Workflow

First, we know the drosophila time series (`dslevin2016dmel`) has imprecise chronological ages, so we'll build a reference with the `dsgraveley2011` data (as in [the second example of the reference-building vignette](RAPToR-refbuilding.html#ex-2)) and stage the drosophila series on it to get accurate developmental timings.


#### *D. melanogaster* on *C. elegans*
We'll filter the data to keep only the orthologs between *D. melanogaster* and  *C. elegans* (and convert FBgn IDs to WBGene IDs).

Then, we'll build a reference with the `dslevin2016cel` data (and stage its samples on it).

Finally, stage the Drosophila samples on it.

#### *C. elegans* on *D. melanogaster*
We'll filter the data to only the orthologs between *D. melanogaster* and  *C. elegans* (and convert WBGene IDs to FBgn IDs).

Then we'll use the reference built with the `dsgraveley2011` to stage the `dslevin2016cel` samples on it.


#### {.tabset}

Code to generate `glist`, `dslevin2016dmel`, `dslevin2016cel` and `dsgraveley2011` :

##### Hide
##### Show

**Note : set the `data_folder` variable to an existing path on your system where you want to store the objects.**

```{r sc2_load_namespaces, eval = F}
data_folder <- "../inst/extdata/"

requireNamespace("wormRef", quietly = T)
requireNamespace("utils", quietly = T)
requireNamespace("GEOquery", quietly = T) # May need to be installed with bioconductor
requireNamespace("Biobase", quietly = T)
```


```{r sc2_c2tpm_func, code = readLines("convert2tpm.R"), echo = TRUE, eval=FALSE}
```

```{r sc2_load_droso_genes, code = readLines("load_droso_genes.R"), echo = TRUE, eval=FALSE}
```


<br>

`glist`

Get list of ortholog genes between *C. elegans* and *D. melanogaster* from @li2014comparison supplementary table 1.
```{r sc2_load_dmel_orthologs, eval = F}
tmp_file <- paste0(data_folder, "dmel_cel_orth.zip")
tmp_fold <- paste0(data_folder, "dmel_cel_orth/")
f_url <- "https://genome.cshlp.org/content/suppl/2014/05/15/gr.170100.113.DC1/Supplemental_Files.zip"

utils::download.file(url = f_url, destfile = tmp_file)
utils::unzip(tmp_file, exdir = tmp_fold)

glist <- read.table(paste0(tmp_fold, "Supplementary\ files/TableS1\ fly-worm\ ortholog\ pairs.txt"), 
                    skip = 1, h=T, sep = "\t", as.is = T, quote = "\"")
colnames(glist) <- c("fb_id", "dmel_name", "cel_id", "cel_name")
glist$wb_id <- wormRef::Cel_genes[match(glist$cel_id, wormRef::Cel_genes$sequence_name), "wb_id"]

save(glist, file = paste0(data_folder, "sc2_glist.RData"), compress = "xz")

file.remove(tmp_file)
unlink(tmp_fold, recursive = T)
rm(tmp_file, tmp_fold, f_url)
```

<br>

`dslevin2016dmel`

```{r sc2_load_dslevin2016dmel, code = readLines("load_dslevin2016dmel.R"), echo = TRUE, eval=FALSE}
```

<br>

`dslevin2016cel`

```{r sc2_load_dslevin2016cel, eval = F}
geo_dslevin2016cel <- "GSE60755"

g_url_dslevin2016cel <- GEOquery::getGEOSuppFiles(geo_dslevin2016cel, makeDirectory = FALSE, fetch_files = FALSE)
g_file_dslevin2016cel <- paste0(data_folder, "dslevin2016cel.txt.gz")
utils::download.file(url = as.character(g_url_dslevin2016cel$url[1]), destfile = g_file_dslevin2016cel)

X_dslevin2016cel <- read.table(gzfile(g_file_dslevin2016cel), h = T, sep = '\t', as.is = T, row.names = 1, comment.char = "")

# filter poor quality samples
cm_dslevin2016cel <- RAPToR::cor.gene_expr(X_dslevin2016cel, X_dslevin2016cel)
f_dslevin2016cel <- which(0.67 > apply(cm_dslevin2016cel, 1, quantile, probs = .99))
X_dslevin2016cel <- X_dslevin2016cel[, -f_dslevin2016cel]

# convert to rpkm & FBgn
X_dslevin2016cel <- RAPToR::format_ids(X_dslevin2016cel, wormRef::Cel_genes, from = "sequence_name", to = "wb_id")
X_dslevin2016cel <- raw2rpkm(X = X_dslevin2016cel, gene.length = wormRef::Cel_genes, id.col = "wb_id", l.col = "transcript_length")


# pheno data
P_dslevin2016cel <- Biobase::pData(GEOquery::getGEO(geo_dslevin2016cel, getGPL = F)[[1]])

# filter relevant fields/samples
P_dslevin2016cel <- P_dslevin2016cel[, c("title", "geo_accession", "time point (minutes after 4-cell):ch1")]
colnames(P_dslevin2016cel)[3] <- "time"
P_dslevin2016cel$title <- as.character(P_dslevin2016cel$title)

P_dslevin2016cel <- P_dslevin2016cel[P_dslevin2016cel$title %in% colnames(X_dslevin2016cel),]

# formatting
P_dslevin2016cel$age <- as.numeric(P_dslevin2016cel$time) / 60
P_dslevin2016cel <- P_dslevin2016cel[order(P_dslevin2016cel$age),]
X_dslevin2016cel <- X_dslevin2016cel[, P_dslevin2016cel$title]

P_dslevin2016cel$title <- gsub('Metazome_CE_timecourse_', '', P_dslevin2016cel$title)
colnames(X_dslevin2016cel) <- P_dslevin2016cel$title

X_dslevin2016cel <- X_dslevin2016cel[,-127] # remove extra outlier
P_dslevin2016cel <- P_dslevin2016cel[-127,]


dslevin2016cel <- list(g = X_dslevin2016cel, p = P_dslevin2016cel)
save(dslevin2016cel, file = paste0(data_folder, "dslevin2016cel.RData"), compress = "xz")

# cleanup
file.remove(g_file_dslevin2016cel)
rm(geo_dslevin2016cel, g_url_dslevin2016cel, g_file_dslevin2016cel, 
   f_dslevin2016cel, cm_dslevin2016cel, X_dslevin2016cel, P_dslevin2016cel)
```

`dsgraveley2011`

```{r load_dsgraveley2011_c, eval = F}
g_url_dsgraveley2011 <- "ftp://ftp.fruitfly.org/pub/download/modencode_expression_scores/Celniker_Drosophila_Annotation_20120616_1428_allsamps_MEAN_gene_expression.csv.gz"
g_file_dsgraveley2011 <- paste0(data_folder, "dsgraveley2011.csv.gz")
utils::download.file(g_url_dsgraveley2011, destfile = g_file_dsgraveley2011)


X_dsgraveley2011 <- read.table(gzfile(g_file_dsgraveley2011), sep = ',', row.names = 1, h = T)

# convert gene ids to FBgn
X_dsgraveley2011 <- RAPToR::format_ids(X_dsgraveley2011, dmel_genes, from = "gene_name", to = "fb_id")

# select embryo time series samples
X_dsgraveley2011 <- X_dsgraveley2011[,1:12]

P_dsgraveley2011 <- data.frame(sname = colnames(X_dsgraveley2011),
                    age = as.numeric(gsub("em(\\d+)\\.\\d+hr", "\\1", colnames(X_dsgraveley2011))),
                    stringsAsFactors = FALSE)

dsgraveley2011 <- list(g = X_dsgraveley2011, p = P_dsgraveley2011)

save(dsgraveley2011, file = paste0(data_folder, "dsgraveley2011.RData"), compress = "xz")

# cleanup
file.remove(g_file_dsgraveley2011)
rm(g_url_dsgraveley2011, g_file_dsgraveley2011, X_dsgraveley2011, P_dsgraveley2011)
```

###  Normalization
```{r sc2_load_ds, include = F, eval = gen_figs}
load("../inst/extdata/dslevin2016cel.RData")
load("../inst/extdata/dslevin2016dmel.RData")
load("../inst/extdata/dsgraveley2011.RData")

load("../inst/extdata/sc2_glist.RData")
```

```{r sc2_qnorm, eval=gen_figs}
dsgraveley2011$g <- limma::normalizeBetweenArrays(dsgraveley2011$g, method = "quantile")
dsgraveley2011$g <- log(dsgraveley2011$g + 1)

dslevin2016dmel$g <- limma::normalizeBetweenArrays(dslevin2016dmel$g, method = "quantile")
dslevin2016dmel$g <- log(dslevin2016dmel$g + 1)

dslevin2016cel$g <- limma::normalizeBetweenArrays(dslevin2016cel$g, method = "quantile")
dslevin2016cel$g <- log(dslevin2016cel$g + 1)

```


```{r sc2_filt, eval=gen_figs}
dslevin2016cel$g_cel <- format_ids(dslevin2016cel$g, glist, from = "wb_id", to = "wb_id")
dslevin2016dmel$g_cel <- format_ids(dslevin2016dmel$g, glist, from = "fb_id", to = "wb_id")
```



### Staging the Drosophila embryo time course

We know from previous use of the `dslevin2016dmel` dataset that its samples have rather imprecise chronological-developmental synchronicity (see the [refbuilding vignette's 2nd example](RAPToR-refbuilding.html#ex-2) for more information). 
We'll first estimate the age of the samples and use that as a basis for the comparison when staging on *C. elegans* data.

```{r sc2_aedmel, eval=gen_figs}
m_grav <- ge_im(X = dsgraveley2011$g, p = dsgraveley2011$p, formula = "X ~ s(age, bs = 'cr')", nc = 4)

n.inter <- 200 
newdat <- data.frame(
  age = seq(min(dsgraveley2011$p$age), max(dsgraveley2011$p$age), l = n.inter)
)
r_grav <- list(interpGE = predict(m_grav, newdata = newdat), time.series = newdat$age)
ae_dmel <- ae(dslevin2016dmel$g, r_grav$interpGE, r_grav$time.series)

dslevin2016dmel$p$ae <- ae_dmel$age.estimates[,1]
```


### *D. melanogaster* staging on *C. elegans*

#### Build a C. elegans embryo reference

As we have components that are rather noisy with this data, we'll only use 5 components to build the reference.
```{r sc2_build_celref, eval=gen_figs}
pca_cel <- stats::prcomp(dslevin2016cel$g_cel, rank = 10)
nc <- 5 

m_cel <- ge_im(X = dslevin2016cel$g_cel, p = dslevin2016cel$p, formula = "X ~ s(age, bs = 'cr')", nc = nc)

n.inter <- 200 # nb of new timepoints
newdat <- data.frame(
  age = seq(min(dslevin2016cel$p$age), max(dslevin2016cel$p$age), l = n.inter)
)

pred_cel_comp <- predict(m_cel, newdata = newdat, as.c = T) # for plotting

r_cel <- list(interpGE = predict(m_cel, newdata = newdat), time.series = newdat$age)
```

```{r sc2_plot_cel_comps, echo = F, fig.width=9, fig.height=6}
show_fig(expr = {

  par(mfrow = c(2,3))
  invisible(sapply(seq_len(6), function(i){
    plot(dslevin2016cel$p$age, pca_cel$rotation[,i], lwd = 2, col = "darkblue",
         xlab = "age", ylab = "PC", main = paste0("PC", i))
  
    # points(dslevin2016cel$p$age, pca_cel$rotation[,i], type = 'l', lty = 2)
    if(i<=nc){
      points(newdat$age, pred_cel_comp[, i], col = "royalblue", type = 'l', lwd = 4)
      box(lwd = 2, col = "royalblue")
    }
    if(i == 1)
      legend("topleft", bty = 'n', legend = c("dslevin2016cel", "pred"),
             pch = c(1, NA), lty = c(NA, 1), col = c(1, "royalblue"), lwd = c(3,4))
  }))
}, fig.width=9, fig.height=6)

```

Check the reference by staging its samples on it :

```{r sc2_ae_cel, eval=gen_figs}
ae_cel_on_cel <- ae(dslevin2016cel$g_cel, r_cel$interpGE, r_cel$time.series)
```

```{r sc2_ae_cel_plot, echo = F, fig.height=8, fig.width=8, out.width = "60%"}
show_fig(expr = {
plot(dslevin2016cel$p$age, ae_cel_on_cel$age.estimates[,1],
     ylab = "Age estimates on C. elegans reference (h past 4C stage)",
     xlab = "Chronological age (h past 4C stage)", 
     main = "Chronological vs estimates (dslevin2016cel on r_cel)",
     lwd = 2, col = "darkblue",
     cex = .8)
}, fig.height=8, fig.width=8)
```


<br>


#### Stage *D. melanogaster* embryos on *C. elegans* reference


```{r sc2_ae_dmel, eval=gen_figs}
ae_dmel_on_cel <- ae(dslevin2016dmel$g_cel, r_cel$interpGE, r_cel$time.series)
```

```{r sc2_ae_dmel_plot, echo = F, fig.height=8, fig.width=8, out.width="80%"}
show_fig(expr = {
par(mar = c(4,4,4,2))
plot(dslevin2016dmel$p$ae, ae_dmel_on_cel$age.estimates[,1], 
     ylab = "Age estimates on C. elegans reference (h past 4C stage)",
     xlab = "D. melanogaster developmental age (h past 4C stage)", 
     main = "Staging D. melanogaster samples on a C. elegans reference\n(dslevin2016dmel on r_cel)",
     lwd = 2, col = "firebrick",
     cex = .8)
}, fig.height=8, fig.width=8)
```

### *C. elegans* staging on *D. melanogaster*

We have already built a *D. melanogaster* with `dsgraveley2011`, which we can use directly here.

```{r sc2_cel_dmel, eval = gen_figs}
dslevin2016cel$g_dmel <- format_ids(dslevin2016cel$g, glist, from = "wb_id", to = "fb_id")
ae_cel_on_dmel <- ae(dslevin2016cel$g_dmel, r_grav$interpGE, r_grav$time.series)
```



```{r sc2_plot_ae_cel_dmel, echo = F, fig.height=8, fig.width=8, out.width="80%"}
show_fig(expr = {
par(mar = c(4,4,4,2))
plot(ae_cel_on_cel$age.estimates[,1], ae_cel_on_dmel$age.estimates[,1], 
     xlab = "C. elegans age (h past 4C stage)", 
     ylab = "Age estimates on D. melanogaster reference (h past 4C stage)", 
     main = "Staging C. elegans samples on a D. melanogaster reference\n(dslevin2016cel on r_grav)", 
     lwd = 2, col = "darkblue",
     cex = .8)
}, fig.height=8, fig.width=8)
```














