```{r, include = FALSE, eval = T}
knitr::opts_knit$set(out.format = "html", header = "")
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%'
)
options(width=100)

gen_figs <- F
figpath <- "../inst/cdoc/RAPToR-showcase_figs/"
if(!file.exists(figpath)){
  dir.create(figpath)
}


library(RAPToR)
library(RColorBrewer)
library(beeswarm)
library(parallel)
# library(vioplot)

library(limma)
requireNamespace("wormRef")
library(stats)

transp <- function(col, a=.5){
  colr <- col2rgb(col)
  return(rgb(colr[1,], colr[2,], colr[3,], a*255, maxColorValue = 255))
}

png_custom <- function(figname, path = "", 
                       fig.width = 7, fig.height = 5, res = 150, ...){
  png(filename = paste0(path, figname, ".png"), 
      width = fig.width, height = fig.height, res = res, units = "in")
}

show_fig <- function(figname = knitr::opts_current$get("label"), expr, path = figpath, ...){
  if(gen_figs){
    png_custom(figname = figname, path = figpath, ...)
    eval(expr = expr)
    dev.off()
  }
  else{
    knitr::include_graphics(paste0(path, figname, ".png"))
  }
}
```


### The data

### Workflow

First, we know the drosophila time series has imprecise chronological ages, so we'll build a reference with the ds3 data (as in the second example of the reference-building vignette) and stage thhe drosophila series on it.

We'll then filter the data to keep only the orthologs between C. elegans and drosophila (and convert FBgn IDs to WBGene IDs).
Then, we'll build a reference with the C. elegans data.
Finally, stage the drosophila samples on it.


#### {.tabset}

Code to generate `dslevin2016dmel`, `dslevin2016cel`, `dsgraveley2011` and `sc2_glist`.

##### Hide
##### Show
```{r sc2_raw2rpkm_func, eval = FALSE}
raw2rpkm <- function(X, gene.length, id.col = 1, l.col='length'){
  # Compute RPKM from raw counts
  if(!all(rownames(X)%in%gene.length[, id.col])){
    stop("Some genes are missing length info !")
  }
  res <- sapply(colnames(X), function(samp){
    pm <- sum(X[,samp])/1e6
    rpkm <- (X[,samp]/pm)/(gene.length[match(rownames(X), gene.length[, id.col]), l.col]/1000)
  })
  rownames(res) <- rownames(X)
  return(res)
}
```

```{r load_dmel_genes, eval = F}
requireNamespace("biomaRt", quietly = TRUE)

mart <- biomaRt::useMart("ensembl", dataset = "dmelanogaster_gene_ensembl")
dmel_genes <- biomaRt::getBM(attributes = c("ensembl_gene_id", 
                                            "ensembl_transcript_id",
                                            "external_gene_name",
                                            "transcript_length"),
                              mart = mart)
colnames(dmel_genes)[1:3] <- c("fb_id", "transcript_id", "gene_name")

dmel_cel_genes <- biomaRt::getBM(attributes = c("ensembl_gene_id",
                                                "celegans_homolog_ensembl_gene"),
                                                mart = mart)

colnames(dmel_cel_genes) <- c("fb_id", "wb_id")

sc2_glist <- list(dmel_genes = dmel_genes, dmel_cel_genes = dmel_cel_genes)
save(sc2_glist, file = "../inst/extdata/sc2_glist.RData", compress = "xz")
rm(mart, dmel_genes, dmel_cel_genes)
```

```{r sc2_load_dmel_orthologs, eval = F}
tmp_file <- "../inst/extdata/dmel_cel_orth.zip"
tmp_fold <- "../inst/extdata/dmel_cel_orth/"
f_url <- "https://genome.cshlp.org/content/suppl/2014/05/15/gr.170100.113.DC1/Supplemental_Files.zip"

utils::download.file(url = f_url, destfile = tmp_file)
utils::unzip(tmp_file, exdir = tmp_fold)

glist <- read.table(paste0(tmp_fold, "Supplementary\ files/TableS1\ fly-worm\ ortholog\ pairs.txt"), 
                    skip = 1, h=T, sep = "\t", as.is = T, quote = "\"")
colnames(glist) <- c("fb_id", "dmel_name", "cel_id", "cel_name")
glist$wb_id <- wormRef::Cel_genes[match(glist$cel_id, wormRef::Cel_genes$sequence_name), "wb_id"]

save(glist, file = "../inst/extdata/sc2_glist2.RData", compress = "xz")

file.remove(tmp_file)
unlink(tmp_fold, recursive = T)
rm(tmp_file, tmp_fold, f_url)
```


```{r sc2_load_dslevin2016dmel, eval = F}
geo_dslevin2016dmel <- "GSE60471"

g_url_dslevin2016dmel <- GEOquery::getGEOSuppFiles(geo_dslevin2016dmel, makeDirectory = FALSE, fetch_files = FALSE)
g_file_dslevin2016dmel <- "../inst/extdata/dslevin2016dmel.txt.gz"
utils::download.file(url = as.character(g_url_dslevin2016dmel$url[3]), destfile = g_file_dslevin2016dmel)

X_dslevin2016dmel <- read.table(gzfile(g_file_dslevin2016dmel), h = T, sep = '\t', as.is = T, row.names = 1, comment.char = "")

# filter poor quality samples
cm_dslevin2016dmel <- RAPToR::cor.gene_expr(X_dslevin2016dmel, X_dslevin2016dmel)
f_dslevin2016dmel <- which(0.6 > apply(cm_dslevin2016dmel, 1, quantile, probs = .99))
X_dslevin2016dmel <- X_dslevin2016dmel[, -f_dslevin2016dmel]

# convert to rpkm & FBgn
X_dslevin2016dmel <- format_ids(X_dslevin2016dmel, sc2_glist$dmel_genes, from = "fb_id", to = "fb_id")
X_dslevin2016dmel <- raw2rpkm(X = X_dslevin2016dmel, gene.length = sc2_glist$dmel_genes, id.col = "fb_id", l.col = "transcript_length")


# pheno data
P_dslevin2016dmel <- Biobase::pData(GEOquery::getGEO(geo_dslevin2016dmel, getGPL = F)[[1]])

# filter relevant fields/samples
P_dslevin2016dmel <- P_dslevin2016dmel[, c("title", "geo_accession", "time (minutes cellularization stage):ch1")]
colnames(P_dslevin2016dmel)[3] <- "time"
P_dslevin2016dmel$title <- as.character(P_dslevin2016dmel$title)

P_dslevin2016dmel <- P_dslevin2016dmel[P_dslevin2016dmel$title %in% colnames(X_dslevin2016dmel),]
X_dslevin2016dmel <- X_dslevin2016dmel[, P_dslevin2016dmel$title]

# formatting
P_dslevin2016dmel$title <- gsub('Metazome_Drosophila_timecourse_', '', P_dslevin2016dmel$title)
colnames(X_dslevin2016dmel) <- P_dslevin2016dmel$title

P_dslevin2016dmel$age <- as.numeric(P_dslevin2016dmel$time) / 60

dslevin2016dmel <- list(g = X_dslevin2016dmel, p = P_dslevin2016dmel)
save(dslevin2016dmel, file = "../inst/extdata/dslevin2016dmel.RData", compress = "xz")

# cleanup
file.remove(g_file_dslevin2016dmel)
rm(geo_dslevin2016dmel, g_url_dslevin2016dmel, g_file_dslevin2016dmel, 
   f_dslevin2016dmel, cm_dslevin2016dmel, X_dslevin2016dmel, P_dslevin2016dmel)
```


```{r sc2_load_dslevin2016cel, eval = F}
geo_dslevin2016cel <- "GSE60755"

g_url_dslevin2016cel <- GEOquery::getGEOSuppFiles(geo_dslevin2016cel, makeDirectory = FALSE, fetch_files = FALSE)
g_file_dslevin2016cel <- "../inst/extdata/dslevin2016cel.txt.gz"
utils::download.file(url = as.character(g_url_dslevin2016cel$url[1]), destfile = g_file_dslevin2016cel)

X_dslevin2016cel <- read.table(gzfile(g_file_dslevin2016cel), h = T, sep = '\t', as.is = T, row.names = 1, comment.char = "")

# filter poor quality samples
cm_dslevin2016cel <- RAPToR::cor.gene_expr(X_dslevin2016cel, X_dslevin2016cel)
f_dslevin2016cel <- which(0.67 > apply(cm_dslevin2016cel, 1, quantile, probs = .99))
X_dslevin2016cel <- X_dslevin2016cel[, -f_dslevin2016cel]

# convert to rpkm & FBgn
X_dslevin2016cel <- format_ids(X_dslevin2016cel, wormRef::Cel_genes, from = "sequence_name", to = "wb_id")
X_dslevin2016cel <- raw2rpkm(X = X_dslevin2016cel, gene.length = wormRef::Cel_genes, id.col = "wb_id", l.col = "transcript_length")


# pheno data
P_dslevin2016cel <- Biobase::pData(GEOquery::getGEO(geo_dslevin2016cel, getGPL = F)[[1]])

# filter relevant fields/samples
P_dslevin2016cel <- P_dslevin2016cel[, c("title", "geo_accession", "time point (minutes after 4-cell):ch1")]
colnames(P_dslevin2016cel)[3] <- "time"
P_dslevin2016cel$title <- as.character(P_dslevin2016cel$title)

P_dslevin2016cel <- P_dslevin2016cel[P_dslevin2016cel$title %in% colnames(X_dslevin2016cel),]

# formatting
P_dslevin2016cel$age <- as.numeric(P_dslevin2016cel$time) / 60
P_dslevin2016cel <- P_dslevin2016cel[order(P_dslevin2016cel$age),]
X_dslevin2016cel <- X_dslevin2016cel[, P_dslevin2016cel$title]

P_dslevin2016cel$title <- gsub('Metazome_CE_timecourse_', '', P_dslevin2016cel$title)
colnames(X_dslevin2016cel) <- P_dslevin2016cel$title



dslevin2016cel <- list(g = X_dslevin2016cel, p = P_dslevin2016cel)
save(dslevin2016cel, file = "../inst/extdata/dslevin2016cel.RData", compress = "xz")

# cleanup
file.remove(g_file_dslevin2016cel)
rm(geo_dslevin2016cel, g_url_dslevin2016cel, g_file_dslevin2016cel, 
   f_dslevin2016cel, cm_dslevin2016cel, X_dslevin2016cel, P_dslevin2016cel)
```

###  Normalization
```{r sc2_load_ds, include = F, eval = gen_figs}
load("../inst/extdata/dslevin2016cel.RData")
load("../inst/extdata/dslevin2016dmel.RData")
load("../inst/extdata/ds3.RData")

load("../inst/extdata/sc2_glist.RData")
load("../inst/extdata/sc2_glist2.RData")

dsgraveley2011 <- ds3
rm(ds3) ; gc()
set.seed((1))
```

```{r sc2_qnorm, eval=gen_figs}
dsgraveley2011$g <- limma::normalizeBetweenArrays(dsgraveley2011$g, method = "quantile")
dsgraveley2011$g <- log(dsgraveley2011$g + 1)

dslevin2016dmel$g <- limma::normalizeBetweenArrays(dslevin2016dmel$g, method = "quantile")
dslevin2016dmel$g <- log(dslevin2016dmel$g + 1)

dslevin2016cel$g <- dslevin2016cel$g[,-127] # remove extra outlier
dslevin2016cel$p <- dslevin2016cel$p[-127,]
dslevin2016cel$g <- limma::normalizeBetweenArrays(dslevin2016cel$g, method = "quantile")
dslevin2016cel$g <- log(dslevin2016cel$g + 1)

```


```{r sc2_filt, eval = gen_figs}
dslevin2016cel$g_cel <- format_ids(dslevin2016cel$g, glist, from = "wb_id", to = "wb_id")
dslevin2016dmel$g_cel <- format_ids(dslevin2016dmel$g, glist, from = "fb_id", to = "wb_id")
```



#### Stage the drosophila time course

See [refbuilding vignette 2nd example](RAPToR-refbuilding.html#ex-2) to see why this step is done.

```{r sc2_aedmel, eval=gen_figs}
m_grav <- ge_im(X = dsgraveley2011$g, p = dsgraveley2011$p, formula = "X ~ s(age, bs = 'cr')", nc = 4)

n.inter <- 200 
newdat <- data.frame(
  age = seq(min(dsgraveley2011$p$age), max(dsgraveley2011$p$age), l = n.inter)
)
r_grav <- list(interpGE = predict(m_grav, newdata = newdat), time.series = newdat$age)
ae_dmel <- ae(dslevin2016dmel$g, r_grav$interpGE, r_grav$time.series)

dslevin2016dmel$p$ae <- ae_dmel$age.estimates[,1]
```

#### Build a C. elegans embryo reference


```{r sc2_build_celref, eval=gen_figs}
pca_cel <- stats::prcomp(dslevin2016cel$g_cel, rank = 20)
nc <- 4#sum(summary(pca_cel)$importance[3,] < .9) +1
nc
#> [1] 9

m_cel <- ge_im(X = dslevin2016cel$g_cel, p = dslevin2016cel$p, formula = "X ~ s(age, bs = 'cr')", nc = nc)

n.inter <- 200 # nb of new timepoints
newdat <- data.frame(
  age = seq(min(dslevin2016cel$p$age), max(dslevin2016cel$p$age), l = n.inter)
)

pred_cel_comp <- predict(m_cel, newdata = newdat, as.c = T) # for plotting

r_cel <- list(interpGE = predict(m_cel, newdata = newdat), time.series = newdat$age)
```

```{r sc2_plot_cel_comps, echo = F, fig.width=12, fig.height=6}
show_fig(expr = {

  par(mfrow = c(2,4))
  invisible(sapply(seq_len(8), function(i){
    plot(dslevin2016cel$p$age, m_cel$pca$rotation[,i], lwd = 2,
         xlab = "age", ylab = "PC", main = paste0("PC", i))
  
    # points(dslevin2016cel$p$age, pca_cel$rotation[,i], type = 'l', lty = 2)
    points(newdat$age, pred_cel_comp[, i], col = "royalblue", type = 'l', lwd = 3)
    if(i == 1)
      legend("topleft", bty = 'n', legend = c("dslevin2016cel", "pred"),
             pch = c(1, NA), lty = c(NA, 1), col = c(1, "royalblue"), lwd = c(3,4))
  }))
}, fig.width=12, fig.height=6)

```


```{r sc2_ae_cel, eval=gen_figs}
ae_cel_on_cel <- ae(dslevin2016cel$g_cel, r_cel$interpGE, r_cel$time.series)
```

```{r sc2_ae_cel_plot, echo = F, fig.height=8, fig.width=8}
plot(dslevin2016cel$p$age, ae_cel_on_cel$age.estimates[,1], cex = .8)
```


```{r sc2_ae_cel, eval=gen_figs}
ae_dmel_on_cel <- ae(dslevin2016dmel$g_cel, r_cel$interpGE, r_cel$time.series)
```

```{r sc2_ae_plot2, echo = F, fig.height=8, fig.width=8}
plot(dslevin2016dmel$p$age, ae_dmel_on_cel$age.estimates[,1], cex = .8)
```


```{r sc2_cel_dmel}
dslevin2016cel$g_dmel <- format_ids(dslevin2016cel$g, glist, from = "wb_id", to = "fb_id")
ae_cel_on_dmel <- ae(dslevin2016cel$g_dmel, r_grav$interpGE, r_grav$time.series)
```



```{r sc2_plot_ae_cel_dmel, echo = F fig.height=6, fig.width=6}
par(mar = c(4,4,4,2))
plot(ae_cel_on_cel$age.estimates[,1], ae_cel_on_dmel$age.estimates[,1], 
     xlab = "C. elegans age", ylab = "Drosophila age estimate", main = "C. elegans sample age estimates\non a Drosophila reference", cex = .8)
```














