---
title: "`wormAge` usage"
output: 
  rmarkdown::html_document :
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{wormAge-usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: ../inst/REFERENCES.bib
author: Romain Bulteau
date: May 2019
---
  
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  out.width = '100%'
)
options(width=100)
quick_build <- F # wether to cache heavy-computation chunks 
```


<br>
<br>
`wormAge` was developped to mediate the issue of developmental differences between samples by estimating the age of samples from their gene expression profiles. 
This is a major problem in the field of *C. elegans*, where many factors can impact development speed. Let's get started.

```{r setup, message=FALSE, results='hide'}
library(wormAge)
```



# How does it work ?

The method works in a three-step process. 

 1. From a reference gene expression time series (several of which are included in this package), a near-continous, **high-temporal-resolution reference** is built.
 1. A **correlation profile** of your samples against this reference is computed from the gene expression information, the peak of which corresponds to the estimated age.
 1. A **bootstrap procedure** of the previous step on random subsets of genes is performed to give a confidence interval of the estimates.



# What data can be used ?
The `wormAge` package allows you to estimate the developmental age of samples from their *gene expression profiles*.
This means that any data providing information on gene expression on a large scale is appropriate : RNA-seq counts (or RPKM), MicroArray, Chips...

<div class="alert alert-danger">
  <strong>Warning :</strong><br/>
**The data must not be gene-centered**, as this destroys the relationship between gene levels within a sample.
</div>

## Converting IDs
The only transformation the data may need is to convert its probe IDs or gene IDs to match the reference series'. All the references included in the package use WormBase Gene IDs (*e.g.* `WBGene00016153`)

To help with this conversion, a gene ID reference table is included in the package, with `WormBase.Gene.ID`, `Public.Name` and `Sequence.Name`, obtained through [WormBase](http://www.wormbase.org).

```{r gene_ids}
data("gene_ids")
head(gene_ids)
```

```{r rm_geneids, echo=F}
rm(gene_ids) 
invisible(gc())
```


One can also use the `biomaRt` Bioconductor package to convert gene IDs, as shown in the example below.
```{r biomart_ex, eval=FALSE}
requireNamespace("biomaRt", quietly = T)

# establish connection with the biomart
mart <- biomaRt::useMart("ensembl", dataset = "celegans_gene_ensembl")

# Perform bm query
gene_ids_bm <- biomaRt::getBM(attributes = c("wormbase_gene",
                                             "ensembl_gene_id",
                                             "external_gene_name"), 
                              filters = "external_gene_name", # return matched gene names
                              values = head(gene_ids$Public.Name), # query IDs
                              mart = mart)
head(gene_ids_bm)
#>    wormbase_gene ensembl_gene_id external_gene_name
#> 1 WBGene00000001  WBGene00000001              aap-1
#> 2 WBGene00000002  WBGene00000002              aat-1
#> 3 WBGene00000003  WBGene00000003              aat-2
#> 4 WBGene00000004  WBGene00000004              aat-3
#> 5 WBGene00000005  WBGene00000005              aat-4
#> 6 WBGene00000006  WBGene00000006              aat-5
```
The biomart holds an extensive amount of different gene or probe IDs, which should be able to cover all needs for conversion. 
Getting these lists can also be done manually via the [biomart website](https://www.ensembl.org/biomart)



## Example loading a dataset

We'll fetch a GEO dataset to use for examples from here on. Time series make for good models since we can easily check if age estimates are correct. 
Let's load [GSE52861](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52861), a dataset from @hendriks2014.

```{r load_hendriks, cache=quick_build}
temp <- tempfile() # make a temporary file for unzipping the data
# The data we need is in the supplementary files of this dataset
download.file(
  "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE52861&format=file&file=GSE52861_ce_geneExpression_polyA_normalized.txt.gz", temp)

geno.hendriks <- read.table(gzfile(temp), h=T, sep = '\t', row.names = 1)
unlink(temp) # close temporary file connection

```

This data is in the form of normalized log counts.
```{r head_hend}
head(geno.hendriks[,1:2])
```


It's also a good idea to get some meta/pheno data. We can use the `GEOquery` package for that.

```{r pheno_hendriks, message=FALSE, cache=quick_build}
requireNamespace("GEOquery", quietly = T)
requireNamespace("Biobase", quietly = T)

# Get pheno data on the worms
geo_hend <- GEOquery::getGEO("GSE52861", getGPL = F)[[1]]
pheno.hendriks <- Biobase::pData(geo_hend)
# Filter the relevant fields
pheno.hendriks <- pheno.hendriks[, c("title", "growth protocol:ch1", 
                                     "strain:ch1", "time in development:ch1")]
# Extract age as hours from the corresponding field
pheno.hendriks$age <- sapply(pheno.hendriks$`time in development:ch1`, function(s){
  as.integer(strsplit(s, split = ' ')[[1]][1])
})

# formatting...
pheno.hendriks$title <- gsub('-', '.', fixed = T, as.character(pheno.hendriks$title))
pheno.hendriks$title <- gsub('hr', 'h', pheno.hendriks$title)
rm(geo_hend)
head(pheno.hendriks[,-1])
```


Finally, let's select the samples that are N2 and non-Dauer from the dataset.
```{r hen_select}
sel <- pheno.hendriks$`growth protocol:ch1`=="Continuous" & # select non-dauer
  pheno.hendriks$`strain:ch1`=="N2" # select N2s

geno.hendriks <- geno.hendriks[,sel]
pheno.hendriks <- pheno.hendriks[sel,]
```


<br>
<br>



# Performing an age estimate

## Choosing a reference dataset

`wormAge` estimates the developmental age of samples based on correlation with reference time series datasets. 
This means you have to select the proper dataset to run your samples against.
Below is a chart showing the developmental stages and the corresponding datasets available in the package (developmental time is according to the larval 20$^\circ$C reference series from @oudenaarden2013).

```{r plot_ref_timelines, fig.height=7, fig.width=9}
plot_ref_timelines()
```


<!-- While the `Cel_YA_adult1` time series is interpolated on together with the `Cel_larval` time series (and thus actually covers the whole larval spectrum), it's advised to use the `Cel_larval` reference alone if your samples are matched before the start of the young adult series (40h).  -->
<!-- This is because we lose a part of the geneset of each reference to get their overlap, which in turns means less genes will overlap with your samples to stage.  -->
<!-- Maximizing the geneset on which the estimate is performed means more robust estimates. -->



### Loading the reference

To get precise estimates, `wormAge` interpolates on the dynamics of gene expression of the reference series.
You can load a reference dataset with optimal parameters using the `prepare_refdata()` function.

Alternatively, you can take a look at the section on [building your own reference series.](#building-your-own-reference)

For our example, the `Cel_YA_adult1` dataset would be appropriate since our samples go from late L3 to young adult. However, since the dataset we just loaded is part of that reference, we'll stage the early samples of the time series on the `Cel_larval` reference to prove our point.
The `n.inter` parameter corresponds to the resolution of the re-built reference. 
In the interest of lightening the computational load for this vignette, I chose a rather small value, but aim over the 500s if you want optimal results.

```{r prep_ref, cache=quick_build}
ref <- prepare_refdata("Cel_larval", n.inter = 200)
```

Note that the age estimates will be given in the time unit and scale of the chosen reference (here, in hours post-hatching at 20$^\circ$C)

We know our example data has worms grown at 25$^\circ$C and that the `Cel_larval` reference goes up to 48 hours post-hatching. We can thus safely stage samples with a chronological time lower than $48/1.5=32$ hours.
```{r sel_tostage}
sel <- which(pheno.hendriks$age < 32)
```


## Performing a simple age estimate

All we need to do now is run the `estimate.worm_age()` function.
```{r echo=F}
set.seed(10)
```

```{r perform_ae, cache=quick_build}
ae_hendriks <- estimate.worm_age(samp = geno.hendriks[,sel], # select the early worms 
                                 refdata = ref$interpol.gene_expr,
                                 ref.time_series = ref$time.series)
```

Let's do a quick check of our estimates against the chronological time points.

Since the staged samples were grown at 25$^\circ$C, and the reference at 20, we can observe the well-known growth speed difference of 1.5 between worms grown at these temperatures by fitting a simple linear model.


```{r hend_age_diff}
lm_hendriks <- lm(ae_hendriks$age.estimates[,1]~pheno.hendriks$age[sel])
summary(lm_hendriks)
```

```{r echo=F, fig.height=4, fig.width=6, out.width="50%"}
plot(ae_hendriks$age.estimates[,1]~pheno.hendriks$age[sel], 
     xlab = "Chronological age (25C)", ylab = "Estimated age (20C)", 
     main = "Chronological vs Estimated ages for Hendriks series")
abline(lm_hendriks, lwd=2, col=2)
legend("bottomright", legend = "lm_hendriks", lwd=3, col=2, bty='n')
```


## Understanding the output
The output of the `estimate.worm_age()` is an `ae` object including various elements such as the age estimate and a 95% confidence interval obtained through booststrapping (age estimate on random gene subsets).

<br>
General information can be accessed via the `summary()` function.
```{r summ_ae}
summary(ae_hendriks)
```



<br>
If you want to retreive the estimates and their confidence intervals, they are accessible through `$age.estimates`.
```{r show_ae}
head(ae_hendriks$age.estimates)
```

The table holds the following :

 - `age.estimate`, the global estimate for the sample (whole gene set).
 - `2.5%`, `97.5%`, the lower and upper bounds of the bootstrapped age estimates' confidence interval.
 - `cor.score`, the correlation score of the global estimate.
 - `IC.imbalance`, ratio of distances between global estimate and bounds of the interval ; if this value is above 5, it usually means that the bootstrap estimates 'jump' between two very different correlation peaks, in which case a warning is given in the `summary`. More information on this is given in the [advanced usage](#if-the-bootstrap-estimates-jump-around) sections

### Plotting

Estimates and confidence intervals can also be displayed in the form of a dotchart with the default plot function.

```{r plot_ae, fig.height=5, fig.width=9, fig.align='center'}
plot(ae_hendriks, main="Age estimates on Hendriks development series")
```


The `ae` object also holds the correlation scores of the samples against the reference series and their 95% intervals accross the bootstrap. 
You can also plot these correlation profiles with `plot_cor.ae()`

```{r plot_cor, fig.height=6, fig.width=9, fig.align='center'}
par(mfrow=c(2,2))
plot_cor.ae(ae_hendriks, subset = 1:4)
```

The 95 % interval is represented for both the estimate (red bars) and the correlation scores along the reference series (black dotted lines). 
The global estimate is also given below the interval.

<br>
<br>


# About reference datasets

## References included in the package
There are currently 3 reference datasets included in the package. 
All are *C. elegans* time series :

 - `Cel_embryo` is an embryo development time series (time in minutes past the 4-cell state) produced by @hashimshony2015, starting at the single cell embryo all the way to hatching.
 - `Cel_larval` holds larval development time series (time in hours post-hatching for 20$^\circ$C, aside from `time.series.25`) from @oudenaarden2013. This data holds 2 series one at 20$^\circ$C and one at 25. The data loaded with `prepare_refdata()` uses both, with the 25$^\circ$C series' time scaled to 20.
 - `Cel_YA_adult1` holds data from early L3 to Adult (time in hours post-hatching for 20$^\circ$C), produced by @sterken2019 and @hendriks2014. 

<!-- - `Cel_YA_adult2` holds data from late L3 to Adult (time in hours post-hatching for 20$^\circ$C), produced by @reinke2004. Due to its age, this data was initially extremely noisy. The data in this object is actually reconstructed from the few developmental signals we were able to pick up by analyzing it together with the data from @oudenaarden2013. -->
 
Sources and links to original data are available in the documentation of the datasets (*e.g.* `?Cel_larval`).


This reference data is the basis for building high-resolution temporal references (through calling the `prepare_refdata()` function). 
To briefly put it, this is achieved by interpolating on *Independant Component Analysis* components showing relevant dynamics with respect to time, and reconstructing the theoretical gene expression data from them. 
This process is detailed in the next section.


## Building your own reference

If you happen to be working with a different strain (or even a different organism), you can build your own reference to stage your samples. The simplicity of the method allows for its wide application range, though we remain dependant on a good time series.
To do this, you will require a time series of gene expression data on your favourite organism as well as some knowledge about Independant Component Analysis (ICA).

<div class="alert alert-info">
  <strong>Why use ICA ?</strong><br/>
As with Principal Component Analysis (PCA), the goal of ICA is to find the loadings of your rows (here, genes) and columns (samples) in a new space, maximizing a given criterion.
However, where PCA will maximize *variance* (effectively performing *dimension reduction*), ICA will maximize the *independence* of components (effectively performing *signal extraction*).

Since gene expression is highly correlated within different groups of genes, ICA is very good at picking up their signals. 
Using a time series of expression data as input, this means that dynamics with respect to time are cleanly extracted by the components.
</div>

The example dataset we loaded earlier will, once again, be used to illustrate the process.

### Performing an ICA on the time series

Before launching an ICA, you must select the number of components you wish to extract. 
There is no exact rule for how to go about this in our particular scenario, so my rule of thumb is to try out increasingly larger values until you start seeing many redundant or unusable components. 
By unusable, I mean that there is no clear dynamic *w.r.t* time we can interpolate on.

The best way to check for this is to plot the ICA components along time.

```{r ica_hend, cache=quick_build}
requireNamespace("ica", quietly = T)

time.series <- pheno.hendriks$age
nc <- 8 # nb of components
ica.hendriks <- ica::icafast(geno.hendriks, nc)
```

You can look at the documentation for `ica::icafast` to see what is returned by the function. 
Our main focus here is the `$M` matrix which contains the samples' loadings (as rows) on the components (as columns).


```{r plot_ica_hend, fig.height=5, fig.width=10}
par(mfrow=c(2,4))
invisible(sapply(1:nc, function(i){
  plot(time.series, ica.hendriks$M[,i], 
       xlab = "time (h)", ylab = "ICA component",
       main = paste("Comp.", i), 
       lwd=2)
}))
```

In the components above, all are suitable for interpolation. 

### Interpolating on selected components
This step corresponds to finding optimal span parameters for the loess regression.
For conciseness' sake, I'll show the process with only one component.

```{r interp_hend_one, fig.height=3, fig.width=9, warning=FALSE}
par(mfrow=c(1,3))

# new time series values
xs <- seq(from = min(time.series), to = max(time.series), 
          length.out = 200)

invisible(sapply(c(.2, .4, .9), function(sp){
  get.spline(ICA = ica.hendriks, comp = 4,
             time.series = time.series, 
             span = sp,
             pred.x = xs, plot = T)}
))
```

As you can see above, using too small a span (left plot) will fit the data "too closely" and sometimes pick up noise along with the trend we want.
Using a span too large (right plot) will completely smooth out the dynamics.

Finding appropriate span values for your selected components is a tedious but necessary process to get the best high-resolution reference possible. 

### Reconstructing theoretical gene expression data
Once the span values are all determined, we need to convert the interpolated components back to gene expression data.
For ICA, we know that by definition $X = SM^T$, with $X$ the input gene expression matrix and $S$ and $M$ being the gene and sample loadings respectively.
Our interpolations over multiple components are essentially an equivalent of $M$, with much higher time resolution. 
All we need to do is the inverse matrix product to get the theoretical gene expression data : $Y = SN^T$, with $N$ the interpolated loadings (rows) of each component (columns).

In `wormAge`, the `interpol_refdata()` function takes care of this process, given $X$, the resolution of the output, the number of ICA components to compute and keep, and the span values for interpolation.

```{r interpol_hend, fig.height=5, fig.width=10, warning=FALSE}
par(mfrow=c(2,4))

sps <- c(.3,.5,.3,.4,.5,.4,.4,.4)
ref_hend <- interpol_refdata(X = geno.hendriks,
                             time.series = time.series,
                             n.inter = 200,  # resolution of interpolation
                             ica.nc = nc,    # nb of components
                             keep.c = 1:8,   # indices of comp. to keep
                             center = TRUE,  # wether to center the ICA (T by default)
                             span = sps,     # span values for each kept component
                             plot = T)
```


And that's it. You can now stage samples on your reference (provided they are within the initial data's timespan).

<br>
<br>

# Advanced usage


## If the bootstrap estimates 'jump around'

In some cases, two different developmental time points of the reference can both be very closely correlated with a sample.
This has the consequence that, with some random gene subsets, the bootstrap estimates will 'jump' between them and the confidence interval will be quite large and uneven (the global estimate will stick to one edge of it).

This can be due to different problems :

 - If the sample is bulk, there can be worms of an unexpected developmental stage contaminating it
 - The overlapping gene set between the sample and the reference may be missing key gene groups necessary to distinguish between larval states ; unlikely, unless 
 - The overlapping gene set is very small (less than a few thousand), which necessarily means that
 - The boostrap gene subset size is too small
 - Random gene subset selection can also in rare cases make the difference between larval stages ambiguous

We can simulate the "small bootstrap set size" by messing with the `bootstrap.set_size` parameter.
```{r boot_jump, cache=quick_build}
set.seed(5) # for reproducible results (get the same random gene subsets each time)

ae_hend_boot <- estimate.worm_age(samp = geno.hendriks[, 4:6], # select a few samples that didn't have the issue
                                  refdata = ref$interpol.gene_expr,
                                  ref.time_series = ref$time.series,
                                  bootstrap.set_size = 500, # make the bt subset size very small
                                  bootstrap.n = 50) # do more bt steps to be sure to get the issue
```

The samples with this problem are detected through the `IC.imbalance` criterion mentioned above (when >5), and marked in the summary by an asterisk.
```{r sum_hen_boot}
summary(ae_hend_boot)
```

We can also show the bootstrap estimates as swarms on the `ae` plot, allowing us to see their distribution along the confidence interval.
```{r pl_hen_boot, fig.height=3, fig.width=8}
plot(ae_hend_boot, show.boot_estimates = TRUE, col.b = 'firebrick')
```

You'll notice that most of our bootstrap estimates are still around the global estimate, and that a few have gone astray 10 hours back (*ie* on the previous molt/larval state, in our case).
If you see this, increasing the bootstrap gene subset size should solve the issue (default is $n/3$, with $n$ the number of overlapping genes between sample and reference).





<br>

If you notice the bootstrap estimates are spread half and half, even with a large number of bootstrap steps (>=100), you may want to consider using a prior for these particular samples. 
Be warned that using priors has a significant impact on the estimate.

```{r rm_boots, echo=F}
rm(ae_hend_boot)
invisible(gc())
```



## Using a prior

In some cases, it may be appropriate to use a prior, helping with the estimate.
In the `estimate.worm_age()` function, priors work by giving the parameters for their gaussian distributions (for each sample). The correlation peaks are then ranked according to the prior's density at their respective times (the correlation profile is unaffected by the prior, only the choice of the correlation peak is).

**This implies that with a prior which is completely off, the estimate may also be wrong ; use with care.**

The priors are given *in the reference series' time scale*, so beware of growth speed difference with temperature. 
For example, the Hendriks series we used above is grown at 25$^\circ$C. 

Due to these possible differences and the bias introduced by the prior, it is recommended to carefully plan its use. 
Performing a first run without priors will give a general idea of the difference between the chronological and developmental age of your samples.

Once the priors are determined, you will need to set the standard deviation of the gaussian centered on the sample with the `prior.params` argument. 
This parameter will also indirectly change the weight of the prior over the correlation score for estimate selection. The recommended value is 10 hours (80 min on the embryonic reference series).

On our example, we can use the ajusted chronological ages for 20$^\circ$C.
```{r hend_prior, cache=quick_build}
priors <- pheno.hendriks$age[sel] * 1.6 - 5 # rough approximation based on our previous lm

ae_hend_prior <- estimate.worm_age(samp = geno.hendriks[,sel],
                                   refdata = ref$interpol.gene_expr,
                                   ref.time_series = ref$time.series,
                                   prior = priors,
                                   prior.params = 10)
```


```{r plot_ae_prior, fig.height=5, fig.width=9, fig.align='center'}
plot(ae_hend_prior, main="Age estimates with priors on Hendriks", 
     show.prior = T, col.p = 'red', l.pos = 'bottomright')

all(ae_hend_prior$age.estimates[,1]==ae_hendriks$age.estimates[,1])
```

As you can see, here the estimates are the same as without the prior. 
Nevertheless, we've fixed the first and third sample's "jumping" bootstrap estimates.


# References
