---
title: "Preparing data for `wormAge` use"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{prepare_data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: refs.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE, results='hide'}
library(wormAge)
requireNamespace("biomaRt", quietly = T)
requireNamespace("GEOquery", quietly = T)
```

## Which data can be used ?
The `wormAge` package allows one to estimate the developmental age of samples from their *gene expression profiles*.
This means that any data providing information on gene expression on a large scale is appropriate : RNA-seq counts (or RPKM), MicroArray, Chips...

**The data must not be gene-centered**, as this destroys the relationship between gene levels within a sample.

The only transformation the data may need is to convert its probe IDs or gene IDs to WormBase Gene IDs

### Converting IDs
To help with this conversion, a gene ID reference table is included in the package, with `WormBase.Gene.ID`, `Public.Name` and `Sequence.Name`, obtained through [WormBase](http://www.wormbase.org).

```{r gene_ids}
data("gene_ids")
head(gene_ids)
```

One can also use the `biomaRt` Bioconductor package to convert gene IDs, as shown in the example below.
```{r biomart_ex, eval=FALSE}
# establish connection with the biomart
mart <- biomaRt::useMart("ensembl", dataset = "celegans_gene_ensembl")

# Perform bm query
gene_ids_bm <- biomaRt::getBM(attributes = c("wormbase_gene",
                                             "ensembl_gene_id",
                                             "external_gene_name"), 
                              filters = "external_gene_name", # return matched gene names
                              values = head(gene_ids$Public.Name), # query IDs
                              mart = mart)
head(gene_ids_bm)
#>    wormbase_gene ensembl_gene_id external_gene_name
#> 1 WBGene00000001  WBGene00000001              aap-1
#> 2 WBGene00000002  WBGene00000002              aat-1
#> 3 WBGene00000003  WBGene00000003              aat-2
#> 4 WBGene00000004  WBGene00000004              aat-3
#> 5 WBGene00000005  WBGene00000005              aat-4
#> 6 WBGene00000006  WBGene00000006              aat-5
```
The biomart holds an extensive amount of different gene or probe IDs, which should be able to cover all needs for conversion. 
Getting these lists can also be done manually via the [biomart website](https://www.ensembl.org/biomart)



## Example loading a dataset

We'll fetch a GEO dataset to use for examples and time series make good references since we can easily check if age estimates are correct. 
Let's load [GSE52861](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52861), a dataset from @hendriks2014extensive.

```{r load_hendriks, eval=FALSE}
temp <- tempfile() # make a temporary file for the data

# The data we want are the raw counts in the supplementary files of this dataset
download.file(
  "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE52861&format=file&file=GSE52861_ce_geneExpression_polyA_rawCounts.txt.gz", temp)

geno.hendriks <- read.table(gzfile(temp), h=T, sep = '\t', row.names = 1)
unlink(temp) # close file connection

```

Now, we want expression data which is better characterized by rpkm values than raw counts. 
To transform our dataset, we need to get the gene lengths which we can also do using the biomart.

```{r load_hendriks2, eval=FALSE}
# get the gene position info from the biomart
mart <- biomaRt::useMart("ensembl", dataset = "celegans_gene_ensembl")
gene_coords <- biomaRt::getBM(attributes = c("wormbase_gene",
                                             "start_position","end_position"), 
                              filters = "wormbase_gene", values = rownames(geno.hendriks), 
                              mart = mart)
# compute size
gene_coords$size <- gene_coords$end_position - gene_coords$start_position
# filter genes we have lengths of
geno.hendriks <- geno.hendriks[gene_coords$wormbase_gene,] 
```

We can now transform the raw counts to rpkm.
```{r load_hendriks3, eval=FALSE}
raw2rpkm <- function(expr.mat, gene.length, l.col='length'){
  # compute rpkms from gene length info
  if(!all(rownames(expr.mat)%in%rownames(gene.length))){
    stop("Some genes are missing length info !")
  }
  sapply(colnames(expr.mat), function(samp){
    pm <- sum(expr.mat[,samp])/1e6
    rpkm <- (expr.mat[,samp]/pm)/(gene.length[rownames(expr.mat), l.col]/1000)
  })
}

rownames(gene_coords) <- gene_coords$wormbase_gene
geno.hendriks <- raw2rpkm(geno.hendriks, gene_coords, l.col = "size")
rownames(geno.hendriks) <- gene_coords$wormbase_gene

rm(gene_coords, mart, temp) # clean up unnecessary variables
```

It's also a good idea to get some meta/pheno data. We can use the `GEOquery` package for that.

```{r pheno_hendriks, eval=FALSE}
# Get pheno data on the worms
geo_hend <- GEOquery::getGEO("GSE52861", getGPL = F)[[1]]
pheno.hendriks <- GEOquery::pData(geo_hend)
pheno.hendriks <- pheno.hendriks[, c("title", "geo_accession", "growth protocol:ch1", 
                                     "strain:ch1", "time in development:ch1")]
# Extract age as hours
pheno.hendriks$age <- sapply(pheno.hendriks$`time in development:ch1`, function(s){
  as.integer(strsplit(s, split = ' ')[[1]][1])
})

# formatting...
pheno.hendriks$title <- gsub('-', '.', fixed = T, as.character(pheno.hendriks$title))
pheno.hendriks$title <- gsub('hr', 'h', pheno.hendriks$title)
rm(geo_hend)
```


<br>
See how to estimate the age of the samples in this data in the [`estimate_age`](estimate_age.html) vignette.

## References
